<!doctype html>
<html class="no-js" lang="en" data-content_root="../">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="../genindex.html" /><link rel="search" title="Search" href="../search.html" /><link rel="next" title="Example of Caenorhabditis elegans dataset (Core)" href="01_Realistic_data.html" /><link rel="prev" title="Tutorials" href="index.html" />

    <!-- Generated with Sphinx 8.1.3 and Furo 2024.08.06 -->
        <title>Example of simulation data - LineageGRN 0.0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?v=354aac6f" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?v=302659d7" />
    <link rel="stylesheet" type="text/css" href="../_static/css/override.css" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  --color-brand-primary: #357473;
  --color-brand-content: #357473;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">LineageGRN 0.0.1 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo" src="../_static/logo.png" alt="Logo"/>
  </div>
  
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Installation.html">Installation</a></li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="index.html">Tutorials</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Tutorials</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2 current current-page"><a class="current reference internal" href="#">Example of simulation data</a></li>
<li class="toctree-l2"><a class="reference internal" href="01_Realistic_data.html">Example of Caenorhabditis elegans dataset (Core)</a></li>
<li class="toctree-l2"><a class="reference internal" href="02_Realistic_data.html">Example of Caenorhabditis elegans dataset (Full)</a></li>
<li class="toctree-l2"><a class="reference internal" href="03_Realistic_data.html">Example of Zebrafish</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../API/index.html">API</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="../_sources/Tutorials/00_Simulation_data.md.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section id="example-of-simulation-data">
<h1>Example of simulation data<a class="headerlink" href="#example-of-simulation-data" title="Link to this heading">¶</a></h1>
<hr class="docutils" />
<p>In this tutorial, we demonstrate the basic usage of <strong>LineageGRN</strong> using synthetic simulated data.</p>
<section id="workflow-overview">
<h2>Workflow Overview<a class="headerlink" href="#workflow-overview" title="Link to this heading">¶</a></h2>
<ol class="arabic simple">
<li><p><strong>Input Data</strong>:<br />
We start by providing a time-scaled cell fate map and initializing a gene regulatory network (GRN) for the progenitor cell cluster at the root node.</p></li>
<li><p><strong>GRN Generation</strong>:<br />
By assigning state transition rates to each regulatory interaction, we generate GRNs for all downstream nodes of the fate map, which serve as the ground truth for subsequent analyses.</p></li>
<li><p><strong>Data Simulation</strong>:</p>
<ul class="simple">
<li><p>Based on the GRNs of the leaf nodes (representing cell clusters of sampled cells), we simulate scATAC-seq data.</p></li>
<li><p>We use the <strong>SERGIO</strong> algorithm to generate scRNA-seq data.</p></li>
</ul>
</li>
<li><p><strong>Validation</strong>:<br />
The simulated data enables us to evaluate the performance of <strong>LineageGRN</strong> in inferring and reconstructing dynamic GRNs across the time-scaled fate map.</p></li>
</ol>
<p>This example highlights how LineageGRN integrates scATAC-seq and scRNA-seq data with lineage tracing to reconstruct dynamic gene regulatory networks.</p>
<p><img alt="" src="../_images/simulation_workflow.png" /></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Importing packages</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">os</span> 
<span class="kn">import</span> <span class="nn">re</span>

<span class="kn">from</span> <span class="nn">lineagegrn.plot.plot_analysis_results</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">lineagegrn.downstream_analysis</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">lineagegrn.utils.constants</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">lineagegrn.utils.synthetic_data_generation</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">lineagegrn.cell_lineage_reconstruction</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">lineagegrn.gene_regulatory_network</span> <span class="kn">import</span><span class="o">*</span>
<span class="kn">from</span> <span class="nn">lineagegrn.utils.metrics</span> <span class="kn">import</span> <span class="o">*</span>
</pre></div>
</div>
</section>
<section id="input-data">
<h2>Input Data<a class="headerlink" href="#input-data" title="Link to this heading">¶</a></h2>
<p>In this example, we directly take the topology structure of a time-scaled fate map <code class="docutils literal notranslate"><span class="pre">fate_map_edge_dict</span></code> with 7 nodes (cell clusters) as input. Besides, we need to set a state transition rate <code class="docutils literal notranslate"><span class="pre">lambda</span></code> to measure the probability of a child GRN inheriting the known regulatory relationship in its parent GRN.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">fate_map_path</span><span class="o">=</span><span class="s1">&#39;examples/data/simulation/input/edge_dict_simulation.csv&#39;</span>
<span class="n">edge_dict</span><span class="o">=</span><span class="n">load_fate_map_topology</span><span class="p">(</span><span class="n">fate_map_path</span><span class="p">)</span>
<span class="n">edge</span><span class="o">=</span><span class="n">parse_edge_dict</span><span class="p">(</span><span class="n">edge_dict</span><span class="p">)</span>
<span class="n">fate_map</span><span class="o">=</span><span class="n">FateMap</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
<span class="n">newick_format</span><span class="o">=</span><span class="n">generate_newick</span><span class="p">(</span><span class="n">fate_map</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="s1">&#39;((T3:0.6,T4:0.6)P1:0.4,(T5:0.9,T6:0.9)P2:0.1)P0:0;&#39;</span>
</pre></div>
</div>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># Rscript</span>
<span class="n">Rscript</span><span class="w"> </span><span class="n">plot_fate_map.R</span><span class="w"> </span><span class="s">&quot;((T3:0.6,T4:0.6)P1:0.4,(T5:0.9,T6:0.9)P2:0.1)P0:0;&quot;</span><span class="w"> </span><span class="m">1.2</span><span class="w"> </span><span class="m">5</span><span class="w"> </span><span class="m">0.3</span><span class="w"> </span><span class="n">fate_map_plot.pdf</span>
</pre></div>
</div>
<p><img alt="" src="../_images/a_fate_map.png" /></p>
</section>
<section id="grn-generation">
<h2>GRN Generation<a class="headerlink" href="#grn-generation" title="Link to this heading">¶</a></h2>
<section id="generate-the-root-node-grn-grn0">
<h3>Generate the root node GRN <code class="docutils literal notranslate"><span class="pre">grn0</span></code><a class="headerlink" href="#generate-the-root-node-grn-grn0" title="Link to this heading">¶</a></h3>
<ul class="simple">
<li><p>Give the number of regulators (100) and target genes (150), and set the path to store the root node GRN</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">output_path</span><span class="o">=</span><span class="s1">&#39;examples/results/simulation/extra_output/ground_truth_grn&#39;</span>
</pre></div>
</div>
<ul class="simple">
<li><p>Run <code class="docutils literal notranslate"><span class="pre">generate_root_grn</span></code> and output a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> named <code class="docutils literal notranslate"><span class="pre">grn0</span></code>, whose rows represent regulators and columns represent target genes. Its elements ​​range from -1 to 1, absolute values ​​represent regulatory strength, and symbols represent activation (+) or repression (-).</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">grn0</span><span class="o">=</span><span class="n">generate_root_grn</span><span class="p">(</span><span class="mi">150</span><span class="p">,</span><span class="mi">100</span><span class="p">,</span><span class="n">output_path</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-plaintext notranslate"><div class="highlight"><pre><span></span>      0        1         2         3         4         5         6         7         8         9      ...       140       141       142       143       144       145       146       147       148       149
150  -0.7318   0.5520    0.0000    0.0000    0.0000    0.0000    0.0000    0.0000    0.0000   -0.7566   ...    0.0000   -0.6687   -0.1849    0.0000    0.0000    0.8071   -0.3010    0.0000    0.2533    0.0000
151   0.0000   0.3824    0.4575    0.4303    0.0000   -0.0109   -0.6965    0.0000    0.0000   -0.5088   ...    0.0000    0.0000    0.0000   -0.1950   -0.2014    0.0000   -0.0074    0.0000   -0.8496    0.0000
152  -0.6151  -0.7520    0.0000   -0.5634    0.0000   -0.0332    0.0000    0.0000    0.0000    0.0000   ...    0.0000    0.0000    0.0000   -0.0441   -0.1915    0.0000    0.1513   -0.0167    0.0000   -0.8906
153   0.0000   0.0000    0.0000    0.0000    0.7510    0.0000    0.0000    0.0000   -0.7391   -0.9003   ...    0.0000    0.0000    0.4218    0.6701    0.0000    0.0000    0.5298   -0.4038    0.0000    0.5000
154   0.5171  -0.5487   -0.5332    0.0000    0.0000    0.9371    0.7675    0.1911    0.6494    0.9113   ...    0.0271   -0.0509    0.0000    0.1592    0.6060    0.0000    0.0000    0.0000   -0.9362   -0.4575
...   ...      ...       ...       ...       ...       ...       ...       ...       ...       ...      ...      ...       ...       ...       ...       ...       ...       ...       ...       ...       ...
245  -0.6043  -0.1336    0.0000    0.0000    0.0000    0.0000    0.0000    0.0000    0.0590    0.0000   ...    0.0054   -0.1922    0.6736    0.0000    0.2150    0.0000    0.0000    0.0000    0.0000    0.0000
</pre></div>
</div>
</section>
<section id="generate-the-descendant-grns-grn1-dotso-grn6">
<h3>Generate the  descendant GRNs <code class="docutils literal notranslate"><span class="pre">grn1</span></code>$\dotso$<code class="docutils literal notranslate"><span class="pre">grn6</span></code><a class="headerlink" href="#generate-the-descendant-grns-grn1-dotso-grn6" title="Link to this heading">¶</a></h3>
<p>This step inputs the path of the dynamic reconfiguration of the gene regulatory network (same topological structure as fate map) and outputs a <code class="docutils literal notranslate"><span class="pre">Dict</span></code>, whose length are the number of nodes (cell clusters) on the fate map. The keys of the <code class="docutils literal notranslate"><span class="pre">Dict</span></code> are <code class="docutils literal notranslate"><span class="pre">node_id</span></code> and each value is the corresponding grn, whose columns represent target genes and rows represent regulators.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">grn_dict_path</span><span class="o">=</span><span class="s1">&#39;examples/data/simulation/additional_input_data/grn_dict.csv&#39;</span>
<span class="n">grn_dict</span><span class="o">=</span><span class="n">load_fate_map_topology</span><span class="p">(</span><span class="n">grn_dict_path</span><span class="p">)</span>
<span class="n">decendent_grns</span><span class="o">=</span><span class="n">generate_descendant_grns</span><span class="p">(</span><span class="mi">7</span><span class="p">,</span><span class="mi">150</span><span class="p">,</span><span class="mi">100</span><span class="p">,</span><span class="n">grn_dict</span><span class="p">,</span><span class="n">output_path</span><span class="p">,</span><span class="n">grn0</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="data-simulation">
<h2>Data Simulation<a class="headerlink" href="#data-simulation" title="Link to this heading">¶</a></h2>
<section id="generate-scatac-seq-data">
<h3>Generate scATAC-seq data<a class="headerlink" href="#generate-scatac-seq-data" title="Link to this heading">¶</a></h3>
<p>Run <code class="docutils literal notranslate"><span class="pre">generate_atac_data</span></code> and output a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> named <code class="docutils literal notranslate"><span class="pre">atac_data</span></code> which must include four columns representing <code class="docutils literal notranslate"><span class="pre">target_gene_id</span></code>, <code class="docutils literal notranslate"><span class="pre">regulator_id</span></code>, <code class="docutils literal notranslate"><span class="pre">co-accessibility</span> <span class="pre">score</span></code>, and <code class="docutils literal notranslate"><span class="pre">node_id</span></code> respectively.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">output_path</span><span class="o">=</span><span class="s1">&#39;../data/simulation_data/test/input&#39;</span>
<span class="n">atac_data</span><span class="o">=</span><span class="n">generate_atac_data</span><span class="p">(</span><span class="n">decendent_grns</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">150</span><span class="p">,</span><span class="mi">100</span><span class="p">,</span><span class="n">output_path</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="generate-scrna-seq-data-using-sergio">
<h3>Generate scRNA-seq data using SERGIO<a class="headerlink" href="#generate-scrna-seq-data-using-sergio" title="Link to this heading">¶</a></h3>
<ul class="simple">
<li><p>Generate two files required by SERGIO.</p></li>
<li><p>Run <code class="docutils literal notranslate"><span class="pre">generate_expression_data</span></code> and output a dataframe for each node, whose columns represent sampled cells and rows represent regulator genes and target genes. The element ​​represent the gene expression value.</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">sergio_files_path</span><span class="o">=</span><span class="s1">&#39;examples/results/simulation/extra_output/sergio_files&#39;</span>
<span class="n">output_path</span><span class="o">=</span><span class="s1">&#39;examples/results/simulation/extra_output/expression_files&#39;</span>
<span class="n">generate_expression_data</span><span class="p">(</span><span class="n">decendent_grns</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">150</span><span class="p">,</span><span class="mi">100</span><span class="p">,</span><span class="mi">300</span><span class="p">,</span><span class="n">sergio_files_path</span><span class="p">,</span><span class="n">output_path</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-plaintext notranslate"><div class="highlight"><pre><span></span>Start simulating new level
There are 100 genes to simulate in this layer
Done with current level
Start simulating new level
There are 150 genes to simulate in this layer
Done with current level
Start simulating new level
There are 100 genes to simulate in this layer
Done with current level
Start simulating new level
There are 150 genes to simulate in this layer
Done with current level
Start simulating new level
There are 100 genes to simulate in this layer
Done with current level
Start simulating new level
There are 150 genes to simulate in this layer
Done with current level
Start simulating new level
There are 100 genes to simulate in this layer
Done with current level
Start simulating new level
There are 150 genes to simulate in this layer
Done with current level
</pre></div>
</div>
<ul class="simple">
<li><p>Run <code class="docutils literal notranslate"><span class="pre">convert_expression_file</span></code> to convert gene expression data into LineageGRN input format. The final algorithm input is a dataframe which includes five columns: <code class="docutils literal notranslate"><span class="pre">gene_id</span></code>, <code class="docutils literal notranslate"><span class="pre">cell_id</span></code>, <code class="docutils literal notranslate"><span class="pre">gene</span> <span class="pre">expression_value</span></code>, <code class="docutils literal notranslate"><span class="pre">node_id</span></code>, <code class="docutils literal notranslate"><span class="pre">gene_type</span></code> (regulator or target gene).</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">input_file</span><span class="o">=</span><span class="s1">&#39;examples/results/simulation/extra_output/expression_files&#39;</span>
<span class="n">output_file</span><span class="o">=</span><span class="s1">&#39;examples/data/simulation/input/expression_data.csv&#39;</span>

<span class="n">target_genes_name</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;examples/data/simulation/additional_input_data/target_genes_name.csv&#39;</span><span class="p">)[</span><span class="s1">&#39;x&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">regulatory_genes_name</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;examples/data/simulation/additional_input_data/regulatory_genes_name.csv&#39;</span><span class="p">)[</span><span class="s1">&#39;x&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">convert_expression_file</span><span class="p">(</span><span class="mi">150</span><span class="p">,</span><span class="n">target_genes_name</span><span class="p">,</span><span class="n">regulatory_genes_name</span><span class="p">,</span><span class="n">input_file</span><span class="p">,</span><span class="n">output_file</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-plaintext notranslate"><div class="highlight"><pre><span></span>gene             cell       value   node_id   gene_type
0    target_0       cell_0    10.0    grn3      target_gene
1    target_0       cell_1    18.0    grn3      target_gene
2    target_0       cell_2    15.0    grn3      target_gene
3    target_0       cell_3    13.0    grn3      target_gene
4    target_0       cell_4    11.0    grn3      target_gene
...  ...            ...       ...     ...       ...
74995 regulator_99  cell_295   1.0    grn6      regulator_gene
74996 regulator_99  cell_296   1.0    grn6      regulator_gene
74997 regulator_99  cell_297   1.0    grn6      regulator_gene
74998 regulator_99  cell_298   4.0    grn6      regulator_gene
74999 regulator_99  cell_299   0.0    grn6      regulator_gene

300000 rows × 5 columns
</pre></div>
</div>
</section>
</section>
<section id="id1">
<h2>Data Simulation<a class="headerlink" href="#id1" title="Link to this heading">¶</a></h2>
<section id="input-synthetic-gene-expression-and-co-accessibility-profiles">
<h3>Input synthetic gene expression and co-accessibility profiles<a class="headerlink" href="#input-synthetic-gene-expression-and-co-accessibility-profiles" title="Link to this heading">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">atac_file_path</span> <span class="o">=</span> <span class="s2">&quot;examples/data/simulation/input/atac_data.csv&quot;</span>
<span class="n">expression_file_path</span> <span class="o">=</span> <span class="s2">&quot;examples/data/simulation/input/expression_data.csv&quot;</span>
</pre></div>
</div>
</section>
<section id="run-the-inference-function">
<h3>Run the inference function<a class="headerlink" href="#run-the-inference-function" title="Link to this heading">¶</a></h3>
<p>This step outputs a dataframe for each target gene, whose number of rows is equal to <code class="docutils literal notranslate"><span class="pre">nodes_number</span></code> (7), and the number of columns is related to <code class="docutils literal notranslate"><span class="pre">regulator_number</span></code>. The first column of this dataframe is <code class="docutils literal notranslate"><span class="pre">target_gene_id</span></code>, the second column is <code class="docutils literal notranslate"><span class="pre">node_id</span></code>, and the third to the last column indicates the regulatory strength and regulation mode (activation or repression) of the target gene by the corresponding regulator.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">saved_dir</span><span class="o">=</span><span class="s1">&#39;examples/results/simulation/inferred&#39;</span>
<span class="n">grns</span><span class="o">=</span><span class="n">FateMap</span><span class="p">(</span><span class="n">parse_edge_dict</span><span class="p">(</span><span class="n">grn_dict</span><span class="p">))</span>
<span class="n">grn_inference</span> <span class="o">=</span> <span class="n">GRNInference</span><span class="p">(</span><span class="n">atac_file_path</span><span class="p">,</span><span class="n">expression_file_path</span><span class="p">,</span><span class="n">grns</span><span class="p">,</span><span class="n">saved_dir</span><span class="p">)</span>
<span class="n">grn_inference</span><span class="o">.</span><span class="n">estimate_all_target_genes</span><span class="p">()</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>-2024-09-24 11:32:58,551-INFO-52272 Start fitting target_gene_id:1
-2024-09-24 11:34:56,762-INFO-25168 Finish inferencing leaves grn value for target_gene_id:1
-2024-09-24 11:34:56,763-INFO-25168 Saved grn values for target_gene_id:1
... ...
</pre></div>
</div>
<p>We can also output the target gene network for each target gene in a <code class="docutils literal notranslate"><span class="pre">Dict</span></code> by running the method <code class="docutils literal notranslate"><span class="pre">.get_target_networks</span></code>. In this method, we need to set a threshold (0.1) to screen for the effective regulatory relationships.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">target_networks_dict</span><span class="o">=</span><span class="n">grn_inference</span><span class="o">.</span><span class="n">get_target_networks</span><span class="p">(</span><span class="mf">0.1</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-plaintext notranslate"><div class="highlight"><pre><span></span>{&quot;grn3&quot;: {&quot;0&quot;: {&quot;lambda&quot;: 0.6116428512359448,&quot;grn_value&quot;: [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, -0.4425698207514982, 0.29298224636154474, 0.0, 0.0, 0.4353533375160287, 0.0, 0.5065959649468524, 0.3509382546110441, ..., 0.0, 0.35099408963364565, 0.0, -0.21043651910705435, 0.0]}}
</pre></div>
</div>
<p>The third optional gene regulatory network output format is a <code class="docutils literal notranslate"><span class="pre">Dict</span></code>, where the key of the <code class="docutils literal notranslate"><span class="pre">Dict</span></code> is <code class="docutils literal notranslate"><span class="pre">node_id</span></code>, and each value is a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> with the number of rows being the number of target genes and the number of columns being the number of regulated genes.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">dynamic_networks_dict</span><span class="o">=</span><span class="n">get_dynamic_networks</span><span class="p">(</span><span class="n">saved_dir</span><span class="p">,</span><span class="n">grns</span><span class="p">,</span><span class="mf">0.1</span><span class="p">,</span><span class="n">regulatory_genes_name</span><span class="p">,</span><span class="n">target_genes_name</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>{&#39;grn0&#39;:             regulator_0  regulator_1  regulator_2  regulator_3  regulator_4  \
target_0      -0.379122     0.711633     0.000000     0.000000     0.131632   
target_1      -0.336749     0.193371     0.000000    -0.502988    -0.169468   
target_2      -0.105994     0.000000    -0.180427     0.000000     0.562621   
target_3       0.000000     0.000000     0.000000     0.000000     0.148689   
target_4       0.000000     0.000000     0.258889     0.400358     0.000000   
...                 ...          ...          ...          ...          ...   
target_145    -0.412486     0.000000    -0.246471     0.000000     0.462467   
target_146     0.415713    -0.140728    -0.724423     0.195241     0.000000   
target_147     0.000000    -0.140601     0.000000    -0.461557     0.384506   
target_148     0.000000    -0.244510     0.000000    -0.155289     0.000000   
target_149    -0.234581     0.198741    -0.189440     0.000000     0.000000   

            regulator_5  regulator_6  regulator_7  regulator_8  regulator_9  \
target_0      -0.153831    -0.379208     0.000000     0.000000    -0.253205   
target_1       0.000000     0.000000     0.145741     0.000000     0.000000   
target_2       0.000000    -0.426593     0.287834     0.664276     0.000000   
target_3       0.269684     0.295407     0.540570     0.219219     0.347215   
target_4       0.476580    -0.234921    -0.274365     0.000000     0.000000   
...                 ...          ...          ...          ...          ...   
target_145     0.586070     0.000000     0.000000     0.115736     0.565307   
target_146     0.000000     0.118883     0.285013     0.219476    -0.100838   
target_147     0.138593     0.000000    -0.411954    -0.144207    -0.209958   
target_148     0.000000     0.000000    -0.143893     0.000000     0.294248   
target_149     0.000000    -0.515080     0.000000     0.000000    -0.385077   
...
target_147            -0.157546  0.000000  0.193603  
target_148             0.716554  0.790531 -0.183264  
[150 rows x 100 columns]}
</pre></div>
</div>
<section id="visualization">
<h4>Visualization<a class="headerlink" href="#visualization" title="Link to this heading">¶</a></h4>
<p>For a fate map, we can use the <code class="docutils literal notranslate"><span class="pre">.get_path</span></code> method to output the path of any specified leaf node and visualize the inferred gene regulatory network along the path.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">grns</span><span class="o">.</span><span class="n">get_path</span><span class="p">(</span><span class="s1">&#39;grn5&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="s1">&#39;grn0&#39;</span><span class="p">,</span> <span class="s1">&#39;grn2&#39;</span><span class="p">,</span> <span class="s1">&#39;grn5&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>Take the gene regulatory network in the selected path from <code class="docutils literal notranslate"><span class="pre">dynamic_network_dict</span></code> and transform it in format. Use the <code class="docutils literal notranslate"><span class="pre">get_gene_interaction</span></code> function to output a three-column <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>, where the first column is <code class="docutils literal notranslate"><span class="pre">regulator_id</span></code>, the second column is <code class="docutils literal notranslate"><span class="pre">target_gene_id</span></code>, and the third column is <code class="docutils literal notranslate"><span class="pre">regulatory_strength</span></code>.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">network0</span><span class="p">,</span><span class="n">network2</span><span class="p">,</span><span class="n">network5</span><span class="o">=</span><span class="n">dynamic_networks_dict</span><span class="p">[</span><span class="s1">&#39;grn0&#39;</span><span class="p">],</span><span class="n">dynamic_networks_dict</span><span class="p">[</span><span class="s1">&#39;grn2&#39;</span><span class="p">],</span><span class="n">dynamic_networks_dict</span><span class="p">[</span><span class="s1">&#39;grn5&#39;</span><span class="p">]</span>
<span class="n">network0</span><span class="o">.</span><span class="n">columns</span><span class="o">=</span><span class="n">network2</span><span class="o">.</span><span class="n">columns</span><span class="o">=</span><span class="n">network5</span><span class="o">.</span><span class="n">columns</span><span class="o">=</span><span class="n">regulatory_genes_name</span>
<span class="n">network0</span><span class="o">.</span><span class="n">index</span><span class="o">=</span><span class="n">network2</span><span class="o">.</span><span class="n">index</span><span class="o">=</span><span class="n">network5</span><span class="o">.</span><span class="n">index</span><span class="o">=</span><span class="n">target_genes_name</span>

<span class="n">gene0_interaction</span><span class="p">,</span> <span class="n">gene2_interaction</span><span class="p">,</span> <span class="n">gene5_interaction</span> <span class="o">=</span> <span class="n">get_gene_interaction</span><span class="p">(</span><span class="n">network0</span><span class="p">),</span><span class="n">get_gene_interaction</span><span class="p">(</span><span class="n">network2</span><span class="p">),</span><span class="n">get_gene_interaction</span><span class="p">(</span><span class="n">network5</span><span class="p">)</span>
<span class="n">gene0_interaction</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;examples/results/simulation/extra_output/gene0_interaction.csv&#39;</span><span class="p">,</span><span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="n">gene2_interaction</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;examples/results/simulation/extra_output/gene2_interaction.csv&#39;</span><span class="p">,</span><span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="n">gene5_interaction</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;examples/results/simulation/extra_output/gene5_interaction.csv&#39;</span><span class="p">,</span><span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
<p>Visualize gene regulatory networks on selected pathways. The gene regulatory networks shown are <code class="docutils literal notranslate"><span class="pre">grn0</span></code>, <code class="docutils literal notranslate"><span class="pre">grn2</span></code> and <code class="docutils literal notranslate"><span class="pre">grn5</span></code>.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># Rscript</span>
<span class="n">Rscript</span><span class="w"> </span><span class="n">plot_grn.R</span><span class="w"> </span><span class="n">examples</span><span class="o">/</span><span class="n">results</span><span class="o">/</span><span class="n">simulation</span><span class="o">/</span><span class="n">extra_output</span><span class="o">/</span><span class="n">expression_files</span><span class="o">/</span><span class="n">Exp_grn5.csv</span><span class="w"> </span><span class="n">examples</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">simulation</span><span class="o">/</span><span class="n">additional_input_data</span><span class="o">/</span><span class="n">target_genes_name.csv</span><span class="w"> </span><span class="n">examples</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">simulation</span><span class="o">/</span><span class="n">additional_input_data</span><span class="o">/</span><span class="n">regulatory_genes_name.csv</span><span class="w"> </span><span class="n">examples</span><span class="o">/</span><span class="n">results</span><span class="o">/</span><span class="n">simulation</span><span class="o">/</span><span class="n">extra_output</span><span class="o">/</span><span class="n">gene0_interaction.csv</span><span class="w"> </span><span class="n">grn0_plot.pdf</span>
<span class="n">Rscript</span><span class="w"> </span><span class="n">plot_grn.R</span><span class="w"> </span><span class="n">examples</span><span class="o">/</span><span class="n">results</span><span class="o">/</span><span class="n">simulation</span><span class="o">/</span><span class="n">extra_output</span><span class="o">/</span><span class="n">expression_files</span><span class="o">/</span><span class="n">Exp_grn5.csv</span><span class="w"> </span><span class="n">examples</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">simulation</span><span class="o">/</span><span class="n">additional_input_data</span><span class="o">/</span><span class="n">target_genes_name.csv</span><span class="w"> </span><span class="n">examples</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">simulation</span><span class="o">/</span><span class="n">additional_input_data</span><span class="o">/</span><span class="n">regulatory_genes_name.csv</span><span class="w"> </span><span class="n">examples</span><span class="o">/</span><span class="n">results</span><span class="o">/</span><span class="n">simulation</span><span class="o">/</span><span class="n">extra_output</span><span class="o">/</span><span class="n">gene2_interaction.csv</span><span class="w"> </span><span class="n">grn2_plot.pdf</span>
<span class="n">Rscript</span><span class="w"> </span><span class="n">plot_grn.R</span><span class="w"> </span><span class="n">examples</span><span class="o">/</span><span class="n">results</span><span class="o">/</span><span class="n">simulation</span><span class="o">/</span><span class="n">extra_output</span><span class="o">/</span><span class="n">expression_files</span><span class="o">/</span><span class="n">Exp_grn5.csv</span><span class="w"> </span><span class="n">examples</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">simulation</span><span class="o">/</span><span class="n">additional_input_data</span><span class="o">/</span><span class="n">target_genes_name.csv</span><span class="w"> </span><span class="n">examples</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">simulation</span><span class="o">/</span><span class="n">additional_input_data</span><span class="o">/</span><span class="n">regulatory_genes_name.csv</span><span class="w"> </span><span class="n">examples</span><span class="o">/</span><span class="n">results</span><span class="o">/</span><span class="n">simulation</span><span class="o">/</span><span class="n">extra_output</span><span class="o">/</span><span class="n">gene5_interaction.csv</span><span class="w"> </span><span class="n">grn5_plot.pdf</span>
</pre></div>
</div>
<p><img alt="" src="../_images/grns.png" /></p>
</section>
</section>
</section>
<section id="validation">
<h2>Validation<a class="headerlink" href="#validation" title="Link to this heading">¶</a></h2>
<section id="calculate-the-auroc-and-auprc-value">
<h3>Calculate the AUROC and AUPRC value.<a class="headerlink" href="#calculate-the-auroc-and-auprc-value" title="Link to this heading">¶</a></h3>
<p>This step traverses all fate map nodes and outputs the AUROC and AUPRC obtained by comparing the inferred gene regulatory network with the real gene regulatory network.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">nodes</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">fate_map</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
<span class="n">auroc_list</span><span class="o">=</span><span class="p">[]</span>
<span class="n">auprc_list</span><span class="o">=</span><span class="p">[]</span>

<span class="k">for</span> <span class="n">node_id</span> <span class="ow">in</span> <span class="n">nodes</span><span class="p">:</span>
    <span class="n">grn_infer</span><span class="o">=</span><span class="n">dynamic_networks_dict</span><span class="p">[</span><span class="n">node_id</span><span class="p">]</span>
    <span class="n">grn_true</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;examples/data/simulation/extra_output/ground_truth_grn/&#39;</span><span class="o">+</span><span class="n">node_id</span><span class="o">+</span><span class="s1">&#39;.csv&#39;</span><span class="p">)</span>
    <span class="n">grn_true</span><span class="o">=</span><span class="n">grn_true</span><span class="o">.</span><span class="n">T</span>
    <span class="n">auroc</span><span class="o">=</span><span class="n">compute_auroc</span><span class="p">(</span><span class="n">grn_infer</span><span class="p">,</span><span class="n">grn_true</span><span class="p">,</span><span class="mf">0.1</span><span class="p">)</span>
    <span class="n">auprc</span><span class="o">=</span><span class="n">compute_auprc</span><span class="p">(</span><span class="n">grn_infer</span><span class="p">,</span><span class="n">grn_true</span><span class="p">,</span><span class="mf">0.1</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">([</span><span class="n">node_id</span><span class="p">,</span><span class="n">auroc</span><span class="p">,</span><span class="n">auprc</span><span class="p">])</span>
    <span class="n">auroc_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">node_id</span><span class="p">,</span><span class="n">auroc</span><span class="p">])</span>
    <span class="n">auprc_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">node_id</span><span class="p">,</span><span class="n">auprc</span><span class="p">])</span>
    
<span class="n">auroc_df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">auroc_list</span><span class="p">)</span>
<span class="n">auroc_df</span><span class="o">.</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;node_id&#39;</span><span class="p">,</span><span class="s1">&#39;value&#39;</span><span class="p">]</span>
<span class="n">auprc_df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">auprc_list</span><span class="p">)</span>
<span class="n">auprc_df</span><span class="o">.</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;node_id&#39;</span><span class="p">,</span><span class="s1">&#39;value&#39;</span><span class="p">]</span>
<span class="n">auprc_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;examples/results/simulation/evaluation_result/auprc.csv&#39;</span><span class="p">,</span><span class="n">index</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="n">auroc_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;examples/results/simulation/evaluation_result/auroc.csv&#39;</span><span class="p">,</span><span class="n">index</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>[&#39;grn0&#39;, 0.5994519770589117, 0.6717228828294495]
[&#39;grn1&#39;, 0.6709248585954783, 0.712408224107377]
[&#39;grn2&#39;, 0.6072614237064434, 0.6640750302853287]
[&#39;grn3&#39;, 0.8188240085184858, 0.8332435186071212]
[&#39;grn4&#39;, 0.8177177242199248, 0.8321948886133066]
[&#39;grn5&#39;, 0.8214236542051112, 0.8364298746683348]
[&#39;grn6&#39;, 0.8170608500573491, 0.8317090516001567]
</pre></div>
</div>
<section id="id2">
<h4>Visualization<a class="headerlink" href="#id2" title="Link to this heading">¶</a></h4>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># Rscript</span>
<span class="n">Rscript</span><span class="w"> </span><span class="n">plot_AUC.R</span><span class="w"> </span><span class="s">&quot;((grn3:0.6,grn4:0.6)grn1:0.4,(grn5:0.9,grn6:0.9)grn2:0.1)grn0:0;&quot;</span><span class="w"> </span><span class="n">examples</span><span class="o">/</span><span class="n">results</span><span class="o">/</span><span class="n">simulation</span><span class="o">/</span><span class="n">evaluation_result</span><span class="o">/</span><span class="n">auroc.csv</span><span class="w"> </span><span class="n">AUROC_plot.pdf</span><span class="w"> </span><span class="kc">TRUE</span>
<span class="n">Rscript</span><span class="w"> </span><span class="n">plot_AUC.R</span><span class="w"> </span><span class="s">&quot;((grn3:0.6,grn4:0.6)grn1:0.4,(grn5:0.9,grn6:0.9)grn2:0.1)grn0:0;&quot;</span><span class="w"> </span><span class="n">examples</span><span class="o">/</span><span class="n">results</span><span class="o">/</span><span class="n">simulation</span><span class="o">/</span><span class="n">evaluation_result</span><span class="o">/</span><span class="n">auprc.csv</span><span class="w"> </span><span class="n">AUPRC_plot.pdf</span><span class="w"> </span><span class="kc">TRUE</span>
</pre></div>
</div>
<p><img alt="" src="../_images/f_AUROC.png" /> <img alt="" src="../_images/e_AUPRC.png" /></p>
</section>
</section>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="01_Realistic_data.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Example of Caenorhabditis elegans dataset (Core)</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="index.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Tutorials</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2024, Xiaochen Yu
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Example of simulation data</a><ul>
<li><a class="reference internal" href="#workflow-overview">Workflow Overview</a></li>
<li><a class="reference internal" href="#input-data">Input Data</a></li>
<li><a class="reference internal" href="#grn-generation">GRN Generation</a><ul>
<li><a class="reference internal" href="#generate-the-root-node-grn-grn0">Generate the root node GRN <code class="docutils literal notranslate"><span class="pre">grn0</span></code></a></li>
<li><a class="reference internal" href="#generate-the-descendant-grns-grn1-dotso-grn6">Generate the  descendant GRNs <code class="docutils literal notranslate"><span class="pre">grn1</span></code>$\dotso$<code class="docutils literal notranslate"><span class="pre">grn6</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#data-simulation">Data Simulation</a><ul>
<li><a class="reference internal" href="#generate-scatac-seq-data">Generate scATAC-seq data</a></li>
<li><a class="reference internal" href="#generate-scrna-seq-data-using-sergio">Generate scRNA-seq data using SERGIO</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id1">Data Simulation</a><ul>
<li><a class="reference internal" href="#input-synthetic-gene-expression-and-co-accessibility-profiles">Input synthetic gene expression and co-accessibility profiles</a></li>
<li><a class="reference internal" href="#run-the-inference-function">Run the inference function</a><ul>
<li><a class="reference internal" href="#visualization">Visualization</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#validation">Validation</a><ul>
<li><a class="reference internal" href="#calculate-the-auroc-and-auprc-value">Calculate the AUROC and AUPRC value.</a><ul>
<li><a class="reference internal" href="#id2">Visualization</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="../_static/documentation_options.js?v=d45e8c67"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/scripts/furo.js?v=5fa4622c"></script>
    </body>
</html>