<!doctype html>
<html class="no-js" lang="en" data-content_root="../">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="../genindex.html" /><link rel="search" title="Search" href="../search.html" /><link rel="next" title="Example of Caenorhabditis elegans dataset (Full)" href="02_Realistic_data.html" /><link rel="prev" title="Example of simulation data" href="00_Simulation_data.html" />

    <!-- Generated with Sphinx 8.1.3 and Furo 2024.08.06 -->
        <title>Example of Caenorhabditis elegans dataset (Core) - LineageGRN 0.0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?v=354aac6f" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?v=302659d7" />
    <link rel="stylesheet" type="text/css" href="../_static/css/override.css" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  --color-brand-primary: #357473;
  --color-brand-content: #357473;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">LineageGRN 0.0.1 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo" src="../_static/logo.png" alt="Logo"/>
  </div>
  
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Installation.html">Installation</a></li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="index.html">Tutorials</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Tutorials</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="00_Simulation_data.html">Example of simulation data</a></li>
<li class="toctree-l2 current current-page"><a class="current reference internal" href="#">Example of Caenorhabditis elegans dataset (Core)</a></li>
<li class="toctree-l2"><a class="reference internal" href="02_Realistic_data.html">Example of Caenorhabditis elegans dataset (Full)</a></li>
<li class="toctree-l2"><a class="reference internal" href="03_Realistic_data.html">Example of Zebrafish</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../API/index.html">API</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="../_sources/Tutorials/01_Realistic_data.md.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section id="example-of-caenorhabditis-elegans-dataset-core">
<h1>Example of Caenorhabditis elegans dataset (Core)<a class="headerlink" href="#example-of-caenorhabditis-elegans-dataset-core" title="Link to this heading">¶</a></h1>
<hr class="docutils" />
<p>In this tutorial, we utilized gene expression data and chromatin co-accessibility data, encompassing 34 regulatory genes and 41 target genes, along with the well-characterized C. elegans cell lineage, to infer gene regulatory networks and visualize their dynamic reconfiguration along the lineage.</p>
<p>This example demonstrates how to identify key regulatory genes driving cell differentiation and distinguish between specific and constitutive regulatory interactions at various developmental stages.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">lineagegrn.cell_lineage_reconstruction</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">lineagegrn.gene_regulatory_network</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">lineagegrn.plot.plot_analysis_results</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">lineagegrn.downstream_analysis</span> <span class="kn">import</span> <span class="o">*</span>
</pre></div>
</div>
<section id="data-preprocessing">
<h2>Data Preprocessing<a class="headerlink" href="#data-preprocessing" title="Link to this heading">¶</a></h2>
<p>This section provides an overview of the preprocessing steps applied to the datasets used in this study. The data includes <strong>scRNA-seq data</strong>, and <strong>scATAC-seq data</strong>, which were processed using standard methods widely adopted in the field. Below, we outline the preprocessing logic and describe the final input formats required for the pipeline.</p>
<section id="atac-seq-data-preprocessing">
<h3>ATAC-seq Data Preprocessing<a class="headerlink" href="#atac-seq-data-preprocessing" title="Link to this heading">¶</a></h3>
<p><strong>Overall Workflow:</strong></p>
<ol class="arabic simple">
<li><p><strong>Input Data</strong>:
-The raw data consists of three files that were processed using the <code class="docutils literal notranslate"><span class="pre">cicero</span></code> package.</p></li>
<li><p><strong>Chromatin Co-accessibility</strong>:
-After processing, two peaks (<code class="docutils literal notranslate"><span class="pre">peak1</span></code> and <code class="docutils literal notranslate"><span class="pre">peak2</span></code>) were identified as chromatin fragments, along with a co-accessibility score (<code class="docutils literal notranslate"><span class="pre">co-access</span></code>). Rows with <code class="docutils literal notranslate"><span class="pre">co-access</span></code> values less than 0.1 were removed.</p></li>
<li><p><strong>Mapping Peaks to Genes</strong>:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">peak2</span></code> fragments were matched to gene IDs.</p></li>
<li><p>The gene IDs were further matched to gene names, which were identified as <strong>target genes</strong>.</p></li>
</ul>
</li>
<li><p><strong>Motif Scanning on <code class="docutils literal notranslate"><span class="pre">peak1</span></code></strong>:</p>
<ul class="simple">
<li><p>Motif scanning was performed on <code class="docutils literal notranslate"><span class="pre">peak1</span></code> regions to identify the transcription factors (TFs) that bind to each peak.</p></li>
<li><p>A single <code class="docutils literal notranslate"><span class="pre">peak1</span></code> may have multiple motif IDs, indicating that a chromatin fragment can bind multiple TFs. Consequently, one target gene may have multiple regulatory genes.</p></li>
</ul>
</li>
<li><p><strong>Final Matching</strong>:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">peak1</span></code> was matched to regulatory genes (TFs), and <code class="docutils literal notranslate"><span class="pre">peak2</span></code> was matched to target genes.</p></li>
<li><p>The final dataset links:</p>
<ul>
<li><p>A single <code class="docutils literal notranslate"><span class="pre">peak1</span></code> to <strong>multiple regulatory genes</strong> (TFs).</p></li>
<li><p>A single <code class="docutils literal notranslate"><span class="pre">peak2</span></code> to <strong>one target gene</strong>.</p></li>
<li><p>The co-accessibility (<code class="docutils literal notranslate"><span class="pre">co-access</span></code>) score represents the shared score between the target gene and its regulatory genes.</p></li>
</ul>
</li>
</ul>
</li>
</ol>
</section>
<section id="scrna-seq-data-preprocessing">
<h3>scRNA-seq Data Preprocessing<a class="headerlink" href="#scrna-seq-data-preprocessing" title="Link to this heading">¶</a></h3>
<p><strong>Overall Workflow:</strong></p>
<ol class="arabic simple">
<li><p><strong>Input Data</strong>:
-The input data for scRNA-seq analysis is typically organized as a gene × cell matrix</p></li>
<li><p><strong>Quality Control</strong>:</p>
<ul class="simple">
<li><p>Cells with low gene expression levels or signs of cell death were removed.</p></li>
</ul>
</li>
<li><p><strong>Highly Variable Genes (Optional)</strong>:</p>
<ul class="simple">
<li><p>Highly variable genes were selected for further analysis to focus on biologically significant variation.</p></li>
</ul>
</li>
<li><p><strong>Normalization</strong>:</p>
<ul class="simple">
<li><p>Gene expression levels were normalized such that the average expression of genes in each cell was standardized to 0, and the variance was standardized to 1.</p></li>
</ul>
</li>
</ol>
</section>
<section id="final-input-formats">
<h3>Final Input Formats<a class="headerlink" href="#final-input-formats" title="Link to this heading">¶</a></h3>
<p>The final inputs for gene regulatory network inference are as follows:</p>
<section id="scatac-seq-data">
<h4>scATAC-seq Data:<a class="headerlink" href="#scatac-seq-data" title="Link to this heading">¶</a></h4>
<ul class="simple">
<li><p><strong>Columns</strong>:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">Regulatory</span> <span class="pre">Gene</span></code>: The name of the regulatory gene associated with <code class="docutils literal notranslate"><span class="pre">peak1</span></code>, derived from motif scanning.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Target</span> <span class="pre">Gene</span></code>: The name of the target gene associated with <code class="docutils literal notranslate"><span class="pre">peak2</span></code>, based on genomic position annotation.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Co-access</span> <span class="pre">Score</span></code>: The co-accessibility score calculated by the <code class="docutils literal notranslate"><span class="pre">cicero</span></code> package, representing the prior regulatory relationship strength.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Cell</span> <span class="pre">Cluster</span> <span class="pre">ID/Cell</span> <span class="pre">Type</span></code>: The cluster ID or cell type to which the cell belongs, providing additional context about the regulatory relationship.</p></li>
</ul>
</li>
<li><p><strong>Example</strong>:</p></li>
</ul>
<div class="table-wrapper colwidths-auto docutils container">
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Regulatory Gene (TF)</p></th>
<th class="head"><p>Target Gene</p></th>
<th class="head"><p>Co-access Score</p></th>
<th class="head"><p>Cell Cluster ID/Type</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>GENE_A</p></td>
<td><p>GENE_D</p></td>
<td><p>0.45</p></td>
<td><p>Cluster_1</p></td>
</tr>
<tr class="row-odd"><td><p>GENE_B</p></td>
<td><p>GENE_D</p></td>
<td><p>0.32</p></td>
<td><p>Cluster_2</p></td>
</tr>
<tr class="row-even"><td><p>GENE_C</p></td>
<td><p>GENE_E</p></td>
<td><p>0.28</p></td>
<td><p>Cluster_3</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="scrna-seq-data">
<h4>scRNA-seq Data:<a class="headerlink" href="#scrna-seq-data" title="Link to this heading">¶</a></h4>
<ul class="simple">
<li><p><strong>Columns</strong>:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">Gene</span> <span class="pre">Name</span></code>: Name of the gene (e.g., GENE_A).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Cell</span> <span class="pre">Name</span></code>: Unique identifier for each cell.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Gene</span> <span class="pre">Expression</span></code>: Normalized gene expression value.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Cell</span> <span class="pre">Cluster</span> <span class="pre">ID/Cell</span> <span class="pre">Type</span></code>: The cluster or cell type label for the cell.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Gene</span> <span class="pre">Type</span></code>: Either <code class="docutils literal notranslate"><span class="pre">Regulatory</span> <span class="pre">Gene</span></code> or <code class="docutils literal notranslate"><span class="pre">Target</span> <span class="pre">Gene</span></code>.</p></li>
</ul>
</li>
<li><p><strong>Example</strong>:</p></li>
</ul>
<div class="table-wrapper colwidths-auto docutils container">
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Gene Name</p></th>
<th class="head"><p>Cell Name</p></th>
<th class="head"><p>Gene Expression</p></th>
<th class="head"><p>Cell Cluster ID/Type</p></th>
<th class="head"><p>Gene Type</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>GENE_D</p></td>
<td><p>Cell_1</p></td>
<td><p>1.23</p></td>
<td><p>Cluster_1</p></td>
<td><p>Target Gene</p></td>
</tr>
<tr class="row-odd"><td><p>GENE_A</p></td>
<td><p>Cell_2</p></td>
<td><p>2.45</p></td>
<td><p>Cluster_2</p></td>
<td><p>Regulatory Gene</p></td>
</tr>
</tbody>
</table>
</div>
</section>
</section>
<section id="references-and-tools-used">
<h3>References and Tools Used<a class="headerlink" href="#references-and-tools-used" title="Link to this heading">¶</a></h3>
<ul class="simple">
<li><p><strong>ATAC-seq Data Preprocessing</strong>:</p>
<ul>
<li><p><a class="reference external" href="https://cole-trapnell-lab.github.io/cicero-release/">Cicero R package</a> for chromatin co-accessibility analysis.</p></li>
<li><p>Motif scanning performed using <a class="reference external" href="https://bioconductor.org/packages/motifmatchr">motifmatchr</a>.</p></li>
</ul>
</li>
<li><p><strong>scRNA-seq Data Preprocessing</strong>:</p>
<ul>
<li><p>Quality control, normalization, and highly variable gene selection performed using <a class="reference external" href="https://scanpy.readthedocs.io/">Scanpy</a> or <a class="reference external" href="https://satijalab.org/seurat/">Seurat</a>.</p></li>
</ul>
</li>
</ul>
</section>
</section>
<hr class="docutils" />
<section id="time-scaled-cell-fate-map-retrieval">
<h2>Time-scaled cell fate map retrieval<a class="headerlink" href="#time-scaled-cell-fate-map-retrieval" title="Link to this heading">¶</a></h2>
<p>In this example, we directly use the well-established cell lineage tree of <em>Caenorhabditis elegans</em>. Six subtrees of the lineage are used as cell fate maps.</p>
<p>Taking the MSE-lineage as an example, we input the topological structure of a cell fate map using the <code class="docutils literal notranslate"><span class="pre">Dict</span></code> data type. The cell fate map can then be generated by instantiating an object of the <code class="docutils literal notranslate"><span class="pre">FateMap</span></code> class.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">fate_map_path</span><span class="o">=</span><span class="s1">&#39;examples/data/C.elegans/C.elegans_Core/input_data/MSE/edge_dict_MSE.csv&#39;</span>
<span class="n">edge_dict</span><span class="o">=</span><span class="n">load_fate_map_topology</span><span class="p">(</span><span class="n">fate_map_path</span><span class="p">)</span>
<span class="n">edge</span><span class="o">=</span><span class="n">parse_edge_dict</span><span class="p">(</span><span class="n">edge_dict</span><span class="p">)</span>
<span class="n">fate_map</span><span class="o">=</span><span class="n">FateMap</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
</pre></div>
</div>
<p><img alt="" src="../_images/fate_map2.png" /></p>
</section>
<hr class="docutils" />
<section id="gene-regulatory-network-inference">
<h2>Gene regulatory network inference<a class="headerlink" href="#gene-regulatory-network-inference" title="Link to this heading">¶</a></h2>
<section id="input-single-cell-gene-expression-and-co-accessibility-profiles">
<h3>Input single-cell gene expression and co-accessibility profiles<a class="headerlink" href="#input-single-cell-gene-expression-and-co-accessibility-profiles" title="Link to this heading">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">atac_file_path</span> <span class="o">=</span> <span class="s2">&quot;examples/data/C.elegans/C.elegans_Core/input_data/MSE/atac_data.csv&quot;</span>
<span class="n">expression_file_path</span> <span class="o">=</span> <span class="s2">&quot;examples/data/C.elegans/C.elegans_Core/input_data/MSE/expression_data.csv&quot;</span>

</pre></div>
</div>
</section>
<section id="run-the-gene-regulatory-network-inference-function">
<h3>Run the gene regulatory network inference function<a class="headerlink" href="#run-the-gene-regulatory-network-inference-function" title="Link to this heading">¶</a></h3>
<p>This step generates a dataframe for each target gene. The dataframe contains a number of rows equal to the number of cell clusters and columns corresponding to the number of regulatory genes.</p>
<ul class="simple">
<li><p>The <strong>first column</strong> represents the name of the target gene.</p></li>
<li><p>The <strong>second column</strong> corresponds to the cell cluster id, indicating the specific node in the regulatory network.</p></li>
<li><p>The <strong>third to the last columns</strong> capture the regulatory relationship between the target gene and its regulators, including:</p>
<ul>
<li><p>The <strong>regulatory strength</strong>, which quantifies the effect of each regulator on the target gene.</p></li>
<li><p>The <strong>regulation mode</strong>, specifying whether the regulation is <strong>activation</strong> or <strong>repression</strong>.</p></li>
</ul>
</li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">saved_dir</span> <span class="o">=</span> <span class="s1">&#39;examples/results/C.elegans/C.elegans_Core/MSE&#39;</span>
<span class="n">regulator_names</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;examples/data/C.elegans/C.elegans_Core/additional_input_data/regulatory_genes_name.csv&#39;</span><span class="p">,</span><span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span><span class="mi">0</span><span class="p">])</span>
<span class="n">target_gene_names</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;examples/data/C.elegans/C.elegans_Core/additional_input_data/target_genes_name.csv&#39;</span><span class="p">,</span><span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span><span class="mi">0</span><span class="p">])</span>

<span class="n">grn_inference</span> <span class="o">=</span> <span class="n">GRNInference</span><span class="p">(</span><span class="n">atac_file_path</span><span class="p">,</span> <span class="n">expression_file_path</span><span class="p">,</span><span class="n">fate_map</span><span class="p">,</span> <span class="n">saved_dir</span><span class="p">)</span>
<span class="n">grn_inference</span><span class="o">.</span><span class="n">estimate_all_target_genes</span><span class="p">()</span>
</pre></div>
</div>
<div class="highlight-plaintext notranslate"><div class="highlight"><pre><span></span>2024-09-24 11:32:58,551 - INFO - 52272 Start fitting target_gene_id:eef-1A.1
2024-09-24 11:34:56,762 - INFO - 25168 Finish inferencing leaves grn value for target_gene_id:eef-1A.1
2024-09-24 11:34:56,763 - INFO - 25168 Saved grn values for target_gene_id:eef-1A.1
...
</pre></div>
</div>
<p>This step generates the target gene network for each target gene and stores it in a <code class="docutils literal notranslate"><span class="pre">Dict</span></code> using the method <code class="docutils literal notranslate"><span class="pre">.get_target_networks</span></code>.</p>
<p>In this method, a regulatory strength threshold (default: 0.1) is applied to filter out ineffective regulatory interactions. Only interactions exceeding this threshold are included in the resulting network, ensuring the biological relevance and computational robustness of the analysis.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">target_networks_dict</span><span class="o">=</span><span class="n">grn_inference</span><span class="o">.</span><span class="n">get_target_networks</span><span class="p">(</span><span class="mf">0.1</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<hr class="docutils" />
<section id="downstream-analysis">
<h2>Downstream analysis<a class="headerlink" href="#downstream-analysis" title="Link to this heading">¶</a></h2>
<section id="revealing-how-inferred-grns-reconfigure-along-cell-lineages">
<h3>Revealing how inferred GRNs reconfigure along cell lineages<a class="headerlink" href="#revealing-how-inferred-grns-reconfigure-along-cell-lineages" title="Link to this heading">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">regulator_dict</span><span class="o">=</span><span class="n">get_regulators</span><span class="p">(</span><span class="s1">&#39;nhr-25&#39;</span><span class="p">,</span> <span class="s1">&#39;Ea-&#39;</span><span class="p">,</span> <span class="n">saved_dir</span><span class="p">,</span> <span class="n">regulator_names</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-plaintext notranslate"><div class="highlight"><pre><span></span>{&#39;ceh-16&#39;: 0.3416499361653389, &#39;ceh-32&#39;: 0.5666916385389806, &#39;nhr-23&#39;: -0.4925718598273363, &#39;nhr-2&#39;: 0.24816905174646298, &#39;ceh-27&#39;: 0.1109726233953044, &#39;mab-5&#39;: -0.5323998868840484, &#39;nob-1&#39;: 0.23886076979862184, &#39;ceh-36&#39;: 0.5233501635329588, &#39;lin-1&#39;: 0.2897392528846144, &#39;eor-1&#39;: -0.4594620129243684, &#39;elt-2&#39;: 0.2976629236568841, &#39;irx-1&#39;: -0.13928695618132647, &#39;pal-1&#39;: 0.2489742931460235, &#39;nhr-79&#39;: -0.254491892554381, &#39;nhr-67&#39;: -0.27121213967865765, &#39;pha-4&#39;: 0.27487150765088963, &#39;nhr-25&#39;: 0.30274213501027825, &#39;dpl-1&#39;: 0.18021229240405684, &#39;hlh-16&#39;: 0.16969014424893344, &#39;elt-1&#39;: 0.36983762199901216, &#39;die-1&#39;: -0.8442123287166449, &#39;mnm-2&#39;: 0.7498769338962882, &#39;ref-2&#39;: 0.3727797808889699, &#39;lsy-27&#39;: -0.1369907430615935, &#39;B0310.2&#39;: -0.6490667615035224, &#39;dmd-4&#39;: -0.6883133898208357, &#39;end-1&#39;: 0.33351354696441643, &#39;end-3&#39;: -0.3033015231479436, &#39;tbx-37&#39;: 0.13312002995799968, &#39;hlh-1&#39;: 0.20481617473191782, &#39;ref-1&#39;: 0.3388880575957087, &#39;egl-5&#39;: -0.32889328111019744, &#39;elt-7&#39;: 0.2227128886808342, &#39;hlh-26&#39;: 0.19039319833013132}
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">target_dict</span><span class="o">=</span><span class="n">get_target_genes</span><span class="p">(</span><span class="s1">&#39;elt-7&#39;</span><span class="p">,</span> <span class="s1">&#39;Ea-&#39;</span><span class="p">,</span> <span class="n">saved_dir</span><span class="p">,</span> <span class="n">regulator_names</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-plaintext notranslate"><div class="highlight"><pre><span></span>{&#39;ama-1&#39;: 0.12514667502715765, &#39;B0310.2&#39;: -0.5690056949430876, &#39;B0336.3&#39;: -0.20087389262169697, &#39;ceh-16&#39;: 0.20728454926555562, &#39;ceh-27&#39;: -0.37076693848880266, &#39;ceh-43&#39;: 0.14481679327711647, &#39;die-1&#39;: 0.3327874528028564, &#39;dpl-1&#39;: 0.9878740137951658, &#39;dpy-31&#39;: 0.12992381853423898, &#39;dpy-7&#39;: 0.32438275516616405, &#39;dve-1&#39;: -0.28369422531295474, &#39;eef-1A.1&#39;: 0.13208107870118843, &#39;elt-1&#39;: 0.1485771897337543, &#39;elt-6&#39;: -0.3187646539838563, &#39;elt-7&#39;: -0.1953150517426333, &#39;F16B12.6&#39;: 0.13044802075213238, &#39;his-72&#39;: -0.1813583591231202, &#39;hlh-1&#39;: 0.2157784109364843, &#39;hmg-1.2&#39;: -0.13887148392408816, &#39;hmg-11&#39;: 0.09813520413087781, &#39;hsp-3&#39;: 0.24737695149461586, &#39;irx-1&#39;: -0.2025020427447795, &#39;isw-1&#39;: 0.05843540247802143, &#39;lin-1&#39;: 0.3000627454486636, &#39;lin-11&#39;: -0.034702965083833526, &#39;lin-13&#39;: 0.07181059724759799, &#39;lin-26&#39;: 0.14910339699633657, &#39;lin-39&#39;: 0.047706651933201984, &#39;lir-2&#39;: -0.006536041410562453, &#39;mab-5&#39;: -0.05103641530899973, &#39;mep-1&#39;: 0.2284986996612302, &#39;mml-1&#39;: -0.13379086742843393, &#39;nhr-23&#39;: 0.2373577819074984, &#39;nhr-25&#39;: 0.2227128886808342, &#39;nhr-49&#39;: 0.0037408606991700503, &#39;nhr-57&#39;: 0.13837017790147818, &#39;nhr-68&#39;: -0.3052208724250098, &#39;nob-1&#39;: 0.10537876339917188, &#39;pbrm-1&#39;: 0.22186185639658834, &#39;sma-9&#39;: 0.1577509253281303, &#39;tps-2&#39;: 0.17392960836027058}
</pre></div>
</div>
<section id="visualization">
<h4>Visualization<a class="headerlink" href="#visualization" title="Link to this heading">¶</a></h4>
<ol class="arabic simple">
<li><p>Visualization of the change in the total number of regulatory interactions in the gene regulatory network along the fate map.</p></li>
</ol>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">output_path</span><span class="o">=</span><span class="s1">&#39;examples/results/C.elegans/C.elegans_Core/figures/&#39;</span>
<span class="n">path</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">fate_map</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
<span class="n">plot_dynamic_edges_number</span><span class="p">(</span><span class="n">dynamic_networks_dict</span><span class="p">,</span><span class="n">path</span><span class="p">,</span><span class="n">output_path</span><span class="p">)</span>
</pre></div>
</div>
<p><img alt="" src="../_images/01_dynamic_edge.png" /></p>
<ol class="arabic simple" start="2">
<li><p>Visualization of the change in the number of target genes regulated by the regulatory gene ‘elt-7’ along the fate map.</p></li>
</ol>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">plot_dynamic_regulator_number</span><span class="p">(</span><span class="n">saved_dir</span><span class="p">,</span> <span class="n">regulator_names</span><span class="p">,</span><span class="s1">&#39;elt-7&#39;</span><span class="p">,</span> <span class="n">path</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span><span class="n">output_path</span><span class="p">)</span>
</pre></div>
</div>
<p><img alt="" src="../_images/02_dynamic_target_num.png" /></p>
<ol class="arabic simple" start="3">
<li><p>Visualization of the change in the number of regulatory genes which regulate the expression of target gene ‘elt-7’ along the fate map.</p></li>
</ol>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">plot_dynamic_target_gene</span><span class="p">(</span><span class="n">saved_dir</span><span class="p">,</span> <span class="n">regulator_names</span><span class="p">,</span> <span class="s1">&#39;elt-7&#39;</span><span class="p">,</span> <span class="n">path</span><span class="p">,</span> <span class="mf">0.05</span><span class="p">,</span><span class="n">output_path</span><span class="p">)</span>
</pre></div>
</div>
<p><img alt="" src="../_images/03_dynamic_regulator_num.png" /></p>
<ol class="arabic simple" start="4">
<li><p>Visualization of the changes in the number of negative regulation of each target gene along the fate map.</p></li>
</ol>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">plot_dynamic_regulator_number_heatmap</span><span class="p">(</span><span class="n">target_gene_names</span><span class="p">,</span> <span class="n">regulator_names</span><span class="p">,</span><span class="s1">&#39;negative&#39;</span><span class="p">,</span><span class="n">saved_dir</span><span class="p">,</span> <span class="n">path</span><span class="p">,</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">output_path</span><span class="p">)</span>
</pre></div>
</div>
<p><img alt="" src="../_images/04_negative_regulator_num.png" /></p>
<ol class="arabic simple" start="5">
<li><p>Visualization of the changes in the number of positive regulation of each target gene along the fate map.</p></li>
</ol>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">plot_dynamic_regulator_number_heatmap</span><span class="p">(</span><span class="n">target_gene_names</span><span class="p">,</span> <span class="n">regulator_names</span><span class="p">,</span><span class="s1">&#39;positive&#39;</span><span class="p">,</span><span class="n">saved_dir</span><span class="p">,</span> <span class="n">path</span><span class="p">,</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">output_path</span><span class="p">)</span>
</pre></div>
</div>
<p><img alt="" src="../_images/05_positive_regulator_num.png" /></p>
<ol class="arabic simple" start="6">
<li><p>Visualization of the changes in the total number of regulation of each target gene along the fate map.</p></li>
</ol>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">plot_dynamic_regulator_number_heatmap</span><span class="p">(</span><span class="n">target_gene_names</span><span class="p">,</span> <span class="n">regulator_names</span><span class="p">,</span><span class="s1">&#39;total&#39;</span><span class="p">,</span><span class="n">saved_dir</span><span class="p">,</span> <span class="n">path</span><span class="p">,</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">output_path</span><span class="p">)</span>
</pre></div>
</div>
<p><img alt="" src="../_images/06_total_regulator_num.png" /></p>
<ol class="arabic simple" start="7">
<li><p>Visualization of the dynamic regulatory networks of the regulatory gene ‘elt-7’ along the fate map.</p></li>
</ol>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">plot_dynamic_regulatory_network</span><span class="p">(</span><span class="s1">&#39;elt-7&#39;</span><span class="p">,</span><span class="n">dynamic_networks_dict</span><span class="p">,</span><span class="n">path</span><span class="p">,</span><span class="n">output_path</span><span class="p">,</span><span class="mf">0.1</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">16</span><span class="p">,</span><span class="mi">2</span><span class="p">))</span>
</pre></div>
</div>
<p><img alt="" src="../_images/07_dynamic_network.png" /></p>
<ol class="arabic simple" start="8">
<li><p>Visualization of the changes in the regulatory strength of each regulatory gene on the fate map path (with ‘Ea-’ as the leaf node) on the target gene ‘nhr-25’.</p></li>
</ol>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">path</span><span class="o">=</span><span class="n">fate_map</span><span class="o">.</span><span class="n">get_path</span><span class="p">(</span><span class="s2">&quot;Ea-&quot;</span><span class="p">)</span>
<span class="n">plot_dynamic_regulatory_strength</span><span class="p">(</span><span class="n">saved_dir</span><span class="p">,</span><span class="n">path</span><span class="p">,</span><span class="n">regulator_names</span><span class="p">,</span><span class="s1">&#39;nhr-25&#39;</span><span class="p">,</span><span class="n">output_path</span><span class="p">)</span>
</pre></div>
</div>
<p><img alt="" src="../_images/09_regulatory_strength.png" /></p>
<ol class="arabic simple" start="9">
<li><p>Visualization of the comparison results of the activity of the regulatory gene ‘nhr-25’ at different nodes of different fate maps.</p></li>
</ol>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">map_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;examples/data/C.elegans/C.elegans_Core/additional_input_data/fate_map_topology.csv&#39;</span><span class="p">)</span>
<span class="n">edges_dict_Celegans</span> <span class="o">=</span> <span class="p">{}</span>
    
<span class="k">for</span> <span class="n">top_key</span><span class="p">,</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">map_df</span> <span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;Lineage&quot;</span><span class="p">):</span>
    <span class="n">sub_dict</span> <span class="o">=</span> <span class="p">{</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Parent&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">-&gt;</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Child&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Length&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">group</span><span class="o">.</span><span class="n">iterrows</span><span class="p">()}</span>
    <span class="n">edges_dict_Celegans</span><span class="p">[</span><span class="n">top_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">sub_dict</span>

<span class="n">input_path</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">saved_dir</span><span class="p">)</span>
<span class="n">plot_dynamic_regulator_activity</span><span class="p">(</span><span class="n">edges_dict_Celegans</span><span class="p">,</span><span class="n">input_path</span><span class="p">,</span><span class="s1">&#39;nhr-25&#39;</span><span class="p">,</span><span class="n">regulator_names</span><span class="p">,</span><span class="mf">0.35</span><span class="p">,</span><span class="n">output_path</span><span class="p">)</span>
</pre></div>
</div>
<p><img alt="" src="../_images/08_regulator_activity.png" /></p>
</section>
</section>
<section id="identifying-key-regulatory-genes-driving-cell-differentiation">
<h3>Identifying key regulatory genes driving cell differentiation<a class="headerlink" href="#identifying-key-regulatory-genes-driving-cell-differentiation" title="Link to this heading">¶</a></h3>
<p>This step outputs the results of the inferred key regulatory genes that determine the differentiation direction of downstream cells at each progenitor cell cluster (node). It is a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> containing 8 columns: <code class="docutils literal notranslate"><span class="pre">regulator_id</span></code>, <code class="docutils literal notranslate"><span class="pre">positive_regulatory_strength</span></code>, <code class="docutils literal notranslate"><span class="pre">positive_regulation_number</span></code>, <code class="docutils literal notranslate"><span class="pre">negative_regulation_number</span></code>, <code class="docutils literal notranslate"><span class="pre">PRS</span></code>,	<code class="docutils literal notranslate"><span class="pre">PRN</span></code>, and	<code class="docutils literal notranslate"><span class="pre">node_id</span></code>.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">key_regulators_df_concat</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
<span class="k">for</span> <span class="n">lineage</span> <span class="ow">in</span> <span class="n">edges_dict_Celegans</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
    <span class="n">edges</span> <span class="o">=</span> <span class="n">parse_edge_dict</span><span class="p">(</span><span class="n">edges_dict_Celegans</span><span class="p">[</span><span class="n">lineage</span><span class="p">])</span>
    <span class="n">fatemap</span> <span class="o">=</span> <span class="n">FateMap</span><span class="p">(</span><span class="n">edges</span><span class="p">,</span><span class="n">high_expression_genes</span><span class="p">[</span><span class="n">lineage</span><span class="p">])</span>
    <span class="n">nodes</span><span class="o">=</span><span class="n">fatemap</span><span class="o">.</span><span class="n">node_internals</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">+</span><span class="p">[</span><span class="n">fatemap</span><span class="o">.</span><span class="n">node_root</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">nodes</span><span class="p">:</span>
        <span class="n">input_path</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">saved_dir</span><span class="p">),</span><span class="n">lineage</span><span class="p">)</span>
        <span class="n">key_regulators</span><span class="o">=</span><span class="n">find_diff_genes</span><span class="p">(</span><span class="n">input_path</span><span class="p">,</span><span class="n">fatemap</span><span class="p">,</span><span class="mf">0.05</span><span class="p">,</span><span class="n">regulator_names</span><span class="p">,</span><span class="n">target_gene_names</span><span class="p">,</span><span class="n">node</span><span class="p">)</span>
        <span class="n">high_expression_genes_of_ancestor_node</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">key_regulators</span><span class="o">.</span><span class="n">loc</span><span class="p">[((</span><span class="n">key_regulators</span><span class="p">[</span><span class="s1">&#39;PRS&#39;</span><span class="p">]</span><span class="o">&gt;</span><span class="mf">0.2</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">key_regulators</span><span class="p">[</span><span class="s1">&#39;PRN&#39;</span><span class="p">]</span><span class="o">&gt;</span><span class="mf">0.7</span><span class="p">)),</span><span class="s1">&#39;regulator_id&#39;</span><span class="p">])</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span><span class="n">target_gene_names</span><span class="p">))</span>
        <span class="n">high_expression_genes</span><span class="p">[</span><span class="n">lineage</span><span class="p">][</span><span class="n">node</span><span class="p">]</span><span class="o">=</span><span class="n">high_expression_genes_of_ancestor_node</span>
        <span class="n">fatemap</span> <span class="o">=</span> <span class="n">FateMap</span><span class="p">(</span><span class="n">edges</span><span class="p">,</span><span class="n">high_expression_genes</span><span class="p">[</span><span class="n">lineage</span><span class="p">])</span>
        <span class="n">key_regulators_df_concat</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">key_regulators_df_concat</span><span class="p">,</span><span class="n">key_regulators</span><span class="p">])</span>
</pre></div>
</div>
<div class="highlight-plaintext notranslate"><div class="highlight"><pre><span></span>regulator_id	positive_regulatory_strength	positive_regulation_number	negative_regulation_number	PRS	PRN	node_id							
    ceh-16	3.571054	9.0	0.0	0.396784	1.0	E
    ceh-36	3.414377	9.0	0.0	0.379375	1.0	E
    nhr-2	3.340567	8.0	0.0	0.417571	1.0	E
    elt-7	3.424669	7.0	0.0	0.489238	1.0	E
    nhr-79	2.417552	6.0	0.0	0.402925	1.0	E
    ...	...	...	...	...	...	...
    nhr-67	0.034049	4.0	4.0	0.008512	0.5	P2

</pre></div>
</div>
<section id="id1">
<h4>Visualization<a class="headerlink" href="#id1" title="Link to this heading">¶</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">nodes</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">key_regulators_df_concat</span><span class="p">[</span><span class="s1">&#39;node_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">())</span>
<span class="n">plot_diff_genes</span><span class="p">(</span><span class="n">key_regulators_df_concat</span><span class="p">,</span><span class="n">nodes</span><span class="p">,</span><span class="n">regulator_names</span><span class="p">,</span><span class="n">output_path</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mf">2.5</span><span class="p">))</span>
</pre></div>
</div>
<p><img alt="" src="../_images/10_key_regulator.png" /></p>
</section>
</section>
<section id="identifying-specific-and-constitutive-regulatory-interactions">
<h3>Identifying specific and constitutive regulatory interactions.<a class="headerlink" href="#identifying-specific-and-constitutive-regulatory-interactions" title="Link to this heading">¶</a></h3>
<p>This step outputs the clustering results of each regulatory relationship using fuzzy C-means clustering method. X is a <code class="docutils literal notranslate"><span class="pre">pd.DataFrame</span></code>, where each row represents a regulatory relationship, each column represents a node in the fate map, and each element is 0 (the regulatory relationship dose not exist in the node) or 1 (the regulatory relationship exists in the node); <code class="docutils literal notranslate"><span class="pre">centers</span></code> are the centers of the edge clusters; <code class="docutils literal notranslate"><span class="pre">weight_matrix</span></code> is the membership function of each regulatory relationship in each edge cluster.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">X</span><span class="p">,</span><span class="n">centers</span><span class="p">,</span><span class="n">weight_matrix</span><span class="o">=</span><span class="n">run_fuzzy_C_means_clustering</span><span class="p">(</span><span class="n">saved_dir</span><span class="p">,</span><span class="n">fate_map</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">,</span> <span class="n">regulator_names</span><span class="p">,</span> <span class="n">target_gene_names</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">regulator_names</span><span class="p">),</span><span class="nb">len</span><span class="p">(</span><span class="n">target_gene_names</span><span class="p">),</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-plaintext notranslate"><div class="highlight"><pre><span></span>    (      MSE  MS  E  MSp-  MSa-  Ep-  Ea-
    0       1   1  1     1     1    1    1
    1       1   1  1     1     1    1    1
    2       1   1  1     1     1    1    1
    3       1   1  1     1     1    1    1
    4       1   1  1     1     1    1    0
    ...   ...  .. ..   ...   ...  ...  ...
    1389    1   1  1     1     1    1    1
    1390    0   0  0     0     0    0    0
    1391    0   0  0     0     0    0    0
    1392    0   0  0     0     0    0    0
    1393    1   1  1     1     1    1    1
    
    [1394 rows x 7 columns],

            0         1         2         3         4         5         6
    0  0.046258  0.032178  0.117197  0.039124  0.050451  0.043458  0.947314
    1  0.188813  0.045104  0.963920  0.054591  0.113503  0.880305  0.834711
    2  0.990004  0.983410  0.980623  0.963207  0.947537  0.936966  0.961219
    3  0.933149  0.971438  0.075477  0.961343  0.821624  0.095924  0.134773
    4  0.007278  0.008996  0.011875  0.014350  0.015490  0.014834  0.008857,
                    
                    
                        EdgeCluster_1  EdgeCluster_2  EdgeCluster_3  EdgeCluster_4  \
    ceh-16-&gt;eef-1A.1       0.001904       0.003105       0.989338       0.004157   
    ceh-16-&gt;nhr-23         0.001904       0.003105       0.989338       0.004157   
    ceh-16-&gt;elt-1          0.001904       0.003105       0.989338       0.004157   
    ceh-16-&gt;hlh-1          0.001904       0.003105       0.989338       0.004157   
    ...
    hlh-26-&gt;dpl-1          0.998029  
    hlh-26-&gt;mab-5          0.998029  
    hlh-26-&gt;lin-39         0.001495  
    
    [1394 rows x 5 columns])
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">edges_cluster_to_nodes</span><span class="o">=</span><span class="n">map_edge_clusters_to_nodes</span><span class="p">(</span><span class="n">saved_dir</span><span class="p">,</span> <span class="n">fate_map</span><span class="p">,</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">regulator_names</span><span class="p">,</span><span class="n">target_gene_names</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mf">0.3</span><span class="p">,</span><span class="n">X</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">regulator_names</span><span class="p">),</span><span class="nb">len</span><span class="p">(</span><span class="n">target_gene_names</span><span class="p">),</span><span class="n">weight_matrix</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-plaintext notranslate"><div class="highlight"><pre><span></span>EdgeCluster_1	  EdgeCluster_2	  EdgeCluster_3	  EdgeCluster_4	  EdgeCluster_5
    MSE	0.000000	0.208333	1.000000	0.949686	0.000000
    MS	0.000000	0.000000	1.000000	1.000000	0.000000
    E	0.000000	1.000000	1.000000	0.000000	0.000000
    MSp-	0.008929	0.023810	0.936877	0.981132	0.029046
    MSa-	0.071429	0.148810	0.906977	0.792453	0.053942
    Ep-	0.000000	1.000000	0.860465	0.113208	0.000000
    Ea-	1.000000	0.744048	0.930233	0.163522	0.000000
</pre></div>
</div>
<section id="id2">
<h4>Visualization<a class="headerlink" href="#id2" title="Link to this heading">¶</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">plot_edge_cluster_weight</span><span class="p">(</span><span class="n">weight_matrix</span><span class="p">,</span><span class="n">output_path</span><span class="p">)</span>
</pre></div>
</div>
<p><img alt="" src="../_images/11_edge_cluster_weight.png" /></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">plot_edge_cluster_to_nodes</span><span class="p">(</span><span class="n">edges_cluster_to_nodes</span><span class="p">,</span> <span class="n">output_path</span><span class="p">)</span>
</pre></div>
</div>
<p><img alt="" src="../_images/12_edge_cluster_node.png" /></p>
</section>
</section>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="02_Realistic_data.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Example of Caenorhabditis elegans dataset (Full)</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="00_Simulation_data.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Example of simulation data</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2024, Xiaochen Yu
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Example of Caenorhabditis elegans dataset (Core)</a><ul>
<li><a class="reference internal" href="#data-preprocessing">Data Preprocessing</a><ul>
<li><a class="reference internal" href="#atac-seq-data-preprocessing">ATAC-seq Data Preprocessing</a></li>
<li><a class="reference internal" href="#scrna-seq-data-preprocessing">scRNA-seq Data Preprocessing</a></li>
<li><a class="reference internal" href="#final-input-formats">Final Input Formats</a><ul>
<li><a class="reference internal" href="#scatac-seq-data">scATAC-seq Data:</a></li>
<li><a class="reference internal" href="#scrna-seq-data">scRNA-seq Data:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#references-and-tools-used">References and Tools Used</a></li>
</ul>
</li>
<li><a class="reference internal" href="#time-scaled-cell-fate-map-retrieval">Time-scaled cell fate map retrieval</a></li>
<li><a class="reference internal" href="#gene-regulatory-network-inference">Gene regulatory network inference</a><ul>
<li><a class="reference internal" href="#input-single-cell-gene-expression-and-co-accessibility-profiles">Input single-cell gene expression and co-accessibility profiles</a></li>
<li><a class="reference internal" href="#run-the-gene-regulatory-network-inference-function">Run the gene regulatory network inference function</a></li>
</ul>
</li>
<li><a class="reference internal" href="#downstream-analysis">Downstream analysis</a><ul>
<li><a class="reference internal" href="#revealing-how-inferred-grns-reconfigure-along-cell-lineages">Revealing how inferred GRNs reconfigure along cell lineages</a><ul>
<li><a class="reference internal" href="#visualization">Visualization</a></li>
</ul>
</li>
<li><a class="reference internal" href="#identifying-key-regulatory-genes-driving-cell-differentiation">Identifying key regulatory genes driving cell differentiation</a><ul>
<li><a class="reference internal" href="#id1">Visualization</a></li>
</ul>
</li>
<li><a class="reference internal" href="#identifying-specific-and-constitutive-regulatory-interactions">Identifying specific and constitutive regulatory interactions.</a><ul>
<li><a class="reference internal" href="#id2">Visualization</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="../_static/documentation_options.js?v=d45e8c67"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/scripts/furo.js?v=5fa4622c"></script>
    </body>
</html>