<!doctype html>
<html class="no-js" lang="en" data-content_root="../">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="../genindex.html" /><link rel="search" title="Search" href="../search.html" /><link rel="next" title="API" href="../API/index.html" /><link rel="prev" title="Example of Caenorhabditis elegans dataset (Full)" href="02_Realistic_data.html" />

    <!-- Generated with Sphinx 8.1.3 and Furo 2024.08.06 -->
        <title>Example of Zebrafish - LineageGRN 0.0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?v=354aac6f" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?v=302659d7" />
    <link rel="stylesheet" type="text/css" href="../_static/css/override.css" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  --color-brand-primary: #357473;
  --color-brand-content: #357473;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">LineageGRN 0.0.1 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo" src="../_static/logo.png" alt="Logo"/>
  </div>
  
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Installation.html">Installation</a></li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="index.html">Tutorials</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Tutorials</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="00_Simulation_data.html">Example of simulation data</a></li>
<li class="toctree-l2"><a class="reference internal" href="01_Realistic_data.html">Example of Caenorhabditis elegans dataset (Core)</a></li>
<li class="toctree-l2"><a class="reference internal" href="02_Realistic_data.html">Example of Caenorhabditis elegans dataset (Full)</a></li>
<li class="toctree-l2 current current-page"><a class="current reference internal" href="#">Example of Zebrafish</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../API/index.html">API</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="../_sources/Tutorials/03_Realistic_data.md.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section id="example-of-zebrafish">
<h1>Example of Zebrafish<a class="headerlink" href="#example-of-zebrafish" title="Link to this heading">¶</a></h1>
<hr class="docutils" />
<p>In this tutorial, we utilized single-cell lineage tracing, gene expression and chromatin co-accessibility data of 139 regulatory genes and 34 target genes of zebrafish to infer the dynamic gene regulatory networks along zebrafish heart regeneration process and visualize their dynamic reconfiguration along the lineage.</p>
<p>This example demonstrates how to identify key regulatory genes driving cell differentiation and cell fate bias. Inaddition, we identify the specific and constitutive regulatory interactions at various developmental stages.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="kn">from</span> <span class="nn">lineagegrn.utils.metrics</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">lineagegrn.cell_lineage_reconstruction</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">lineagegrn.gene_regulatory_network</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">lineagegrn.plot.plot_analysis_results</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">lineagegrn.downstream_analysis</span> <span class="kn">import</span> <span class="o">*</span>
</pre></div>
</div>
<section id="data-preprocessing">
<h2>Data Preprocessing<a class="headerlink" href="#data-preprocessing" title="Link to this heading">¶</a></h2>
<section id="input-data">
<h3>1. Input Data<a class="headerlink" href="#input-data" title="Link to this heading">¶</a></h3>
<ul class="simple">
<li><p>The input data is a CSV file containing cell barcodes and sequences.</p></li>
<li><p>The sequences are converted to FASTA format for MUSCLE alignment.</p></li>
</ul>
<p><strong>Example</strong>
The CSV file contains columns for unique cell IDs and raw barcodes:</p>
<div class="table-wrapper colwidths-auto docutils container">
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>cell_id</p></th>
<th class="head"><p>raw_barcodes</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Cell_1</p></td>
<td><p>ATGCGTAC…</p></td>
</tr>
<tr class="row-odd"><td><p>Cell_2</p></td>
<td><p>TTACGCTA…</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="sequence-alignment">
<h3>2. Sequence Alignment<a class="headerlink" href="#sequence-alignment" title="Link to this heading">¶</a></h3>
<ul class="simple">
<li><p>Sequences are aligned using MUSCLE to identify similarities and mutations.</p></li>
</ul>
</section>
<section id="data-parsing-and-processing">
<h3>3. Data Parsing and Processing<a class="headerlink" href="#data-parsing-and-processing" title="Link to this heading">¶</a></h3>
<ul class="simple">
<li><p>Aligned sequences are parsed and saved in CSV format.</p></li>
<li><p>The data is sorted by sequence IDs and processed to generate a mutation matrix.</p></li>
</ul>
</section>
<section id="mutation-matrix-generation">
<h3>4. Mutation Matrix Generation<a class="headerlink" href="#mutation-matrix-generation" title="Link to this heading">¶</a></h3>
<ul class="simple">
<li><p>The mutation matrix is created by comparing sequences in defined regions and identifying differences.</p></li>
</ul>
<p><strong>Example</strong>
The mutation matrix shows sequence differences across defined regions:</p>
<div class="table-wrapper colwidths-auto docutils container">
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Sequence ID</p></th>
<th class="head"><p>Region_1</p></th>
<th class="head"><p>Region_2</p></th>
<th class="head"><p>Region_3</p></th>
<th class="head"><p>…</p></th>
<th class="head"><p>Region_n</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Cell_1</p></td>
<td><p>0</p></td>
<td><p>1</p></td>
<td><p>2</p></td>
<td><p>…</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-odd"><td><p>Cell_2</p></td>
<td><p>1</p></td>
<td><p>0</p></td>
<td><p>2</p></td>
<td><p>…</p></td>
<td><p>3</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="example-output-files">
<h3>Example Output Files<a class="headerlink" href="#example-output-files" title="Link to this heading">¶</a></h3>
<div class="table-wrapper colwidths-auto docutils container">
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p><strong>Output File</strong></p></th>
<th class="head"><p><strong>Description</strong></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">output.fasta</span></code></p></td>
<td><p>Converted FASTA file for MUSCLE alignment</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">parse_aligned_sequences.csv</span></code></p></td>
<td><p>CSV file with parsed aligned sequences</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">sort_parse_aligned_sequences.csv</span></code></p></td>
<td><p>Sorted aligned sequence file</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">mutation_matrix.csv</span></code></p></td>
<td><p>Generated mutation matrix file</p></td>
</tr>
</tbody>
</table>
</div>
</section>
</section>
<section id="time-scaled-cell-fate-map-reconstruction">
<h2>Time-scaled cell fate map reconstruction<a class="headerlink" href="#time-scaled-cell-fate-map-reconstruction" title="Link to this heading">¶</a></h2>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">sc_mat</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;examples/data/zebrafish/input_data/barcode_site.csv&#39;</span><span class="p">,</span><span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">))</span>
<span class="n">cell_types</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;examples/data/zebrafish/input_data/cell_types.csv&quot;</span><span class="p">,</span><span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()[</span><span class="s1">&#39;celltypes&#39;</span><span class="p">]</span>
<span class="n">t_S</span><span class="o">=</span><span class="mi">15</span>

<span class="n">edges</span><span class="p">,</span><span class="n">fate_map</span><span class="o">=</span><span class="n">get_fate_map</span><span class="p">(</span><span class="n">sc_mat</span><span class="p">,</span> <span class="n">cell_types</span><span class="p">,</span> <span class="n">t_S</span><span class="p">,</span> <span class="n">beta</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">max_iter</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">tol</span><span class="o">=</span><span class="mf">1e-5</span><span class="p">,</span><span class="n">edge</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="gene-regulatory-network-inference">
<h2>Gene regulatory network inference<a class="headerlink" href="#gene-regulatory-network-inference" title="Link to this heading">¶</a></h2>
<section id="input-single-cell-gene-expression-and-co-accessibility-profiles">
<h3>Input single-cell gene expression and co-accessibility profiles<a class="headerlink" href="#input-single-cell-gene-expression-and-co-accessibility-profiles" title="Link to this heading">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">atac_file_path</span> <span class="o">=</span> <span class="s2">&quot;examples/data/zebrafish/input_data/atac_data.csv&quot;</span>
<span class="n">expression_file_path</span> <span class="o">=</span> <span class="s2">&quot;examples/data/zebrafish/input_data/expression_data.csv&quot;</span>
</pre></div>
</div>
</section>
<section id="run-the-gene-regulatory-network-inference-function">
<h3>Run the gene regulatory network inference function<a class="headerlink" href="#run-the-gene-regulatory-network-inference-function" title="Link to this heading">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">saved_dir</span> <span class="o">=</span> <span class="s1">&#39;examples/results/zebrafish&#39;</span>

<span class="n">regulator_names</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;examples/data/zebrafish/additional_input_data/regulatory_genes_name.csv&quot;</span><span class="p">)[</span><span class="s1">&#39;x&#39;</span><span class="p">])</span>
<span class="n">target_gene_names</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;examples/data/zebrafish/additional_input_data/target_genes_name.csv&quot;</span><span class="p">)[</span><span class="s1">&#39;x&#39;</span><span class="p">])</span>

<span class="n">grn_inference</span> <span class="o">=</span> <span class="n">GRNInference</span><span class="p">(</span><span class="n">atac_file_path</span><span class="p">,</span> <span class="n">expression_file_path</span><span class="p">,</span><span class="n">fate_map</span><span class="p">,</span> <span class="n">saved_dir</span><span class="p">)</span>
<span class="n">grn_inference</span><span class="o">.</span><span class="n">estimate_all_target_genes</span><span class="p">(</span><span class="mi">20</span><span class="p">)</span>
<span class="n">target_networks_dict</span><span class="o">=</span><span class="n">grn_inference</span><span class="o">.</span><span class="n">get_target_networks</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="downstream-analysis">
<h2>Downstream analysis<a class="headerlink" href="#downstream-analysis" title="Link to this heading">¶</a></h2>
<section id="revealing-how-inferred-grns-reconfigure-along-cell-lineages">
<h3>Revealing how inferred GRNs reconfigure along cell lineages<a class="headerlink" href="#revealing-how-inferred-grns-reconfigure-along-cell-lineages" title="Link to this heading">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">regulator_dict</span><span class="o">=</span><span class="n">get_regulators</span><span class="p">(</span><span class="s1">&#39;admb&#39;</span><span class="p">,</span> <span class="s1">&#39;Tcells&#39;</span><span class="p">,</span> <span class="n">saved_dir</span><span class="p">,</span> <span class="n">regulator_names</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-plaintext notranslate"><div class="highlight"><pre><span></span>{&#39;nr4a1&#39;: 0.11825367777901924, &#39;mafbb&#39;: 0.4228665855088654, &#39;en1a&#39;: 0.6564480066299438, &#39;rorc&#39;: 0.0, &#39;nfatc2b&#39;: 0.6564480066299438, &#39;tbx21&#39;: -0.05690246596441495, &#39;phox2a&#39;: 0.6564480066299438, &#39;pax1a&#39;: 0.6564480066299438, &#39;runx3&#39;: 0.15006246468573334, &#39;gfi1ab&#39;: 0.6564480066299438, &#39;elf3&#39;: 0.6564480066299438, &#39;foxp3a&#39;: 0.011441771853181756, &#39;foxi3b&#39;: 0.6564480066299438, &#39;nr1d4a&#39;: 0.0, &#39;foxd3&#39;: 0.6564480066299438, &#39;irx1b&#39;: 0.6564480066299438, &#39;irx6a&#39;: 0.6564480066299438, &#39;jun&#39;: 0.11799363141220853, &#39;fosl1a&#39;: 0.3142731222979524, &#39;egr1&#39;: 0.20470371209130353, &#39;hsf4&#39;: 0.6564480066299438, &#39;tbpl2&#39;: 0.6564480066299438, &#39;tal2&#39;: 0.0, &#39;isl1&#39;: 0.6564480066299438, &#39;nkx2.5&#39;: 0.20591095572917884, &#39;lhx1a&#39;: 0.6564480066299438, &#39;msc&#39;: 0.6564480066299438, &#39;myog&#39;: 0.6564480066299438, &#39;gbx1&#39;: 0.6564480066299438, &#39;uncx4.1&#39;: 0.6564480066299438, &#39;ascl1a&#39;: 0.0, &#39;nfe2&#39;: 0.6564480066299438, &#39;klf2a&#39;: 0.24606150652311023, &#39;atoh1a&#39;: 0.0, &#39;nr5a1a&#39;: 0.0, &#39;cebpb&#39;: 0.11733108755785718, &#39;pax2a&#39;: 0.6564480066299438, &#39;atf3&#39;: 0.11809881535737017, &#39;creb3l3a&#39;: 0.0, &#39;her6&#39;: -0.06831408713485358, &#39;foxi1&#39;: 0.0}
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">target_dict</span><span class="o">=</span><span class="n">get_target_genes</span><span class="p">(</span><span class="s1">&#39;tbx21&#39;</span><span class="p">,</span> <span class="s1">&#39;Tcells&#39;</span><span class="p">,</span> <span class="n">saved_dir</span><span class="p">,</span> <span class="n">regulator_names</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-plaintext notranslate"><div class="highlight"><pre><span></span>{&#39;BX005375.1&#39;: -4.633242599055995e-05, &#39;ugt1b7&#39;: 0.2529791885195795, &#39;aqp1a.1&#39;: 0.737025601278955, &#39;ostn&#39;: 0.4207163038852826, &#39;igfbp1b&#39;: 0.0017850797118275786, &#39;tnfb&#39;: -0.09093491400400154, &#39;asz1&#39;: 3.947985347750165e-07, &#39;dcn&#39;: -0.004753747246812367, &#39;cfd&#39;: -0.07748921820224187, &#39;si_dkey-217l24.1&#39;: -0.01509920087822015, &#39;cxcr4a&#39;: -0.0378512306228885, &#39;nsmfa&#39;: 1.8091607069702714e-05, &#39;CR318588.3&#39;: -0.0282258993159595, &#39;tpm2&#39;: -0.0017816459815551433, &#39;BX897740.1&#39;: 2.0044666629809764e-06, &#39;blf&#39;: -0.04767230441765874, &#39;admb&#39;: -0.05690246596441495, &#39;podxl&#39;: -0.001734048701291507, &#39;f8&#39;: 0.0, &#39;AL929171.1&#39;: -2.6540383868771896e-05, &#39;vcam1b&#39;: -0.005304703632355419, &#39;jun&#39;: -2.7601266249165197e-05, &#39;nlgn2b&#39;: -3.068405735386321e-05, &#39;her9&#39;: -0.003127006382421302, &#39;thbs1b&#39;: 0.0009647495836709402, &#39;fosl1a&#39;: -0.0014129190040687366, &#39;si_ch211-264e16.1&#39;: -1.6457869942911397e-05, &#39;cxcl18b&#39;: 0.13621886372153472, &#39;rbp4&#39;: -0.36247867407755185, &#39;elfn1b&#39;: 0.00011438702960800442, &#39;marcksl1b&#39;: -0.0009142409113317535, &#39;mon1ba&#39;: -4.3403968920740466e-05, &#39;si_ch211-222l21.1&#39;: 0.02575676825170038, &#39;cox7a1&#39;: -0.07475372604189326, &#39;pim1&#39;: -0.011508602286214246, &#39;desma&#39;: -0.041576357456717165, &#39;zgc_193505&#39;: 0.0007180873560352398, &#39;ier2b&#39;: 0.02174653223004613, &#39;mcl1a&#39;: 0.013063183590168705, &#39;s100a10b&#39;: -0.0006377343212278223, &#39;krt15&#39;: -0.0031156868805427727, &#39;apln&#39;: 0.00046948692598344305, &#39;fn1b&#39;: 0.0015811969241520437, &#39;cxcl8a&#39;: 0.0017456590448221838, &#39;gp1bb&#39;: 0.6860146395138957, &#39;hsf4&#39;: 1.672992080801562e-05, &#39;col5a1&#39;: -0.008350685419361516, &#39;zgc_101699&#39;: 0.2568283426854872, &#39;junba&#39;: 0.032504226186033186, &#39;c4&#39;: -0.004556210885012739, &#39;zgc_158868&#39;: 0.0008192537926147509, &#39;atp1a3b&#39;: -0.06202813634657833, &#39;BX120005.1&#39;: 0.04419941960224298, &#39;hspbp1&#39;: -0.027556939626312713, &#39;btg2&#39;: 0.010330613810812, &#39;BX000363.1&#39;: -0.007961956209559572, &#39;CR759878.1&#39;: 3.4584364256234204e-05, &#39;si_ch73-186j5.2&#39;: 1.1228594597287888e-05, &#39;lacc1&#39;: 0.011115105870892348, &#39;junbb&#39;: 0.24998110515097435, &#39;egr1&#39;: -0.0011793886179748989, &#39;tnni4a&#39;: 0.0522262901782596, &#39;xk&#39;: 3.0204721946474364e-05, &#39;sparc&#39;: -0.004487753255040038, &#39;zmat4b&#39;: 2.4908356842685168e-05, &#39;itk&#39;: -0.0532694658455046, &#39;s1pr4&#39;: -0.006356867635340318, &#39;klf2a&#39;: -0.0032715743040377153, &#39;spock3&#39;: -0.0010797206357842655, &#39;rspo1&#39;: 2.9325106708639112e-05, &#39;si_ch211-1a19.3&#39;: 0.035962243147861304, &#39;acyp2&#39;: 0.0, &#39;ctss2.1&#39;: -0.018158764661243377, &#39;tpm4a&#39;: 0.02004788933967247, &#39;cd74a&#39;: 0.03928207675995783, &#39;si_ch211-214p13.9&#39;: 0.06797891659577164, &#39;itga2.2&#39;: 3.963808220532939e-06, &#39;ypel2a&#39;: 5.3302425170750134e-05, &#39;ldb3a&#39;: 0.055583573887739976, &#39;serpinh1b&#39;: -0.0010020282723044597, &#39;zgc_194275&#39;: 7.714064284401884e-05, &#39;igsf5a&#39;: 3.5374537793422442e-06, &#39;scinlb&#39;: -0.03293516883994402, &#39;scpp1&#39;: -0.00011026649699155427, &#39;lpl&#39;: -0.002616008137955797, &#39;lygl1&#39;: -0.00044548884991896817, &#39;cd7al&#39;: 0.00830244301514105, &#39;cyp1b1&#39;: -0.06089733593561483, &#39;mt2&#39;: -0.0036903605064304004, &#39;sez6b&#39;: -0.00012611494693765744, &#39;f3b&#39;: -0.01914353858964715, &#39;zgc_162730&#39;: 0.17600234509827856, &#39;tppp&#39;: 0.0010192434701365666, &#39;acta2&#39;: -0.0026270381673971703, &#39;hsp70l&#39;: 0.010954641303340982, &#39;mb&#39;: 0.012929994148362906, &#39;CR954285.1&#39;: -0.0221873574097361, &#39;myh11a&#39;: -0.02536951696722507, &#39;uchl1&#39;: 0.33693054834398034, &#39;akap12b&#39;: 0.0009212710344499709, &#39;grb10a&#39;: 0.06104321223031403, &#39;dnajb1b&#39;: -0.04076573172557595, &#39;ctsd&#39;: -0.019413559960119363, &#39;CABZ01044764.2&#39;: -0.00013909781813653793, &#39;ctsba&#39;: -0.07022550794531796, &#39;BX005058.1&#39;: 1.2392187140728611e-05, &#39;fgf7&#39;: -9.897818339458238e-05, &#39;tal2&#39;: 2.4228927530041073e-05, &#39;BX901962.4&#39;: 1.749969831488159e-05, &#39;slc8a1a&#39;: -0.00089554323456153, &#39;hspb8&#39;: -0.0003792937094674213, &#39;krt94&#39;: -0.0006166564056375392, &#39;col1a1b&#39;: 0.00042853720539326353, &#39;si_busm1-57f23.1&#39;: 0.027207486472744288, &#39;hspa5&#39;: 0.005099643990921896, &#39;CR855311.1&#39;: 0.24960742323856827, &#39;gadd45ba&#39;: -0.010060484547078316, &#39;ctgfa&#39;: -0.006929624938919741, &#39;fhl1a&#39;: 0.06997724129028973, &#39;TCIM (1 of many)&#39;: -0.001700092046444444, &#39;atp2a2a&#39;: -0.015705752848306845, &#39;il1b&#39;: 0.7303040476006903, &#39;dennd3b&#39;: -1.8730282034801528e-05, &#39;stpg2&#39;: -2.1673523635479636e-05, &#39;pdxp&#39;: 1.702604310429647e-05, &#39;adh8a&#39;: -0.007692104663767129, &#39;cd9b&#39;: -0.0046171434634400405, &#39;aqp8a.1&#39;: -0.06199390282634467, &#39;zgc_158343&#39;: 0.09333489185719923, &#39;mhc1uba&#39;: -0.018881591269891935, &#39;fip1l1a&#39;: 1.8956416788763786e-05, &#39;mfap5&#39;: -0.0004990266560624871}
</pre></div>
</div>
<section id="visualization">
<h4>Visualization<a class="headerlink" href="#visualization" title="Link to this heading">¶</a></h4>
<ol class="arabic simple">
<li><p>Visualization of the change in the total number of regulatory interactions in the gene regulatory network along the fate map.</p></li>
</ol>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">output_path</span><span class="o">=</span><span class="s1">&#39;examples/results/zebrafish/figures/&#39;</span>
<span class="n">path</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">fate_map</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
<span class="n">plot_dynamic_edges_number</span><span class="p">(</span><span class="n">dynamic_networks_dict</span><span class="p">,</span><span class="nb">list</span><span class="p">(</span><span class="n">fate_map</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span><span class="n">output_path</span><span class="p">)</span>
</pre></div>
</div>
<p><img alt="" src="../_images/01_dynamic_edge_zebrafish.png" /></p>
<ol class="arabic simple" start="2">
<li><p>Visualization of the change in the number of target genes regulated by the regulatory gene ‘tbx21’ along the fate map.</p></li>
</ol>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">plot_dynamic_target_gene</span><span class="p">(</span><span class="n">saved_dir</span><span class="p">,</span> <span class="n">regulator_names</span><span class="p">,</span> <span class="s1">&#39;tbx21&#39;</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">fate_map</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="mi">0</span><span class="p">,</span><span class="n">output_path</span><span class="p">)</span>
</pre></div>
</div>
<p><img alt="" src="../_images/02_dynamic_target_num_zebrafish.png" /></p>
<ol class="arabic simple" start="3">
<li><p>Visualization of the changes in the number of negative regulation of each target gene along the fate map.</p></li>
</ol>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">plot_dynamic_regulator_number_heatmap</span><span class="p">(</span><span class="n">target_gene_names</span><span class="p">,</span> <span class="n">regulator_names</span><span class="p">,</span><span class="s1">&#39;negative&#39;</span><span class="p">,</span><span class="n">saved_dir</span><span class="p">,</span><span class="nb">list</span><span class="p">(</span><span class="n">fate_map</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span><span class="mi">0</span><span class="p">,</span> <span class="n">output_path</span><span class="p">)</span>
</pre></div>
</div>
<p><img alt="" src="../_images/04_negative_regulator_num_zebrafish.png" /></p>
<ol class="arabic simple" start="4">
<li><p>Visualization of the changes in the number of positive regulation of each target gene along the fate map.</p></li>
</ol>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">plot_dynamic_regulator_number_heatmap</span><span class="p">(</span><span class="n">target_gene_names</span><span class="p">,</span> <span class="n">regulator_names</span><span class="p">,</span><span class="s1">&#39;positive&#39;</span><span class="p">,</span><span class="n">saved_dir</span><span class="p">,</span><span class="nb">list</span><span class="p">(</span><span class="n">fate_map</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span><span class="mi">0</span><span class="p">,</span> <span class="n">output_path</span><span class="p">)</span>
</pre></div>
</div>
<p><img alt="" src="../_images/05_positive_regulator_num_zebrafish.png" /></p>
<ol class="arabic simple" start="5">
<li><p>Visualization of the dynamic regulatory networks of the regulatory gene ‘klf2a’ along the fate map path (with ‘Endocardium’ as the leaf node).</p></li>
</ol>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">path</span><span class="o">=</span><span class="n">fate_map</span><span class="o">.</span><span class="n">get_path</span><span class="p">(</span><span class="s2">&quot;Endocardium&quot;</span><span class="p">)</span>
<span class="n">plot_dynamic_regulatory_network</span><span class="p">(</span><span class="s1">&#39;klf2a&#39;</span><span class="p">,</span><span class="n">dynamic_networks_dict</span><span class="p">,</span><span class="n">path</span><span class="p">,</span><span class="n">output_path</span><span class="p">,</span><span class="mf">0.12</span><span class="p">)</span>
</pre></div>
</div>
<p><img alt="" src="../_images/07_dynamic_network_zebrafish.png" /></p>
<ol class="arabic simple" start="6">
<li><p>Visualization of the changes in the regulatory strength of each regulatory gene on the fate map path (with ‘Endocardium’ as the leaf node) on the target gene ‘junbb’.</p></li>
</ol>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">plot_dynamic_regulatory_strength</span><span class="p">(</span><span class="n">saved_dir</span><span class="p">,</span> <span class="n">path</span><span class="p">,</span> <span class="n">regulator_names</span><span class="p">,</span> <span class="s1">&#39;junbb&#39;</span><span class="p">,</span> <span class="n">output_path</span><span class="p">)</span>
</pre></div>
</div>
<p><img alt="" src="../_images/09_regulatory_strength_zebrafish.png" /></p>
</section>
</section>
<section id="identifying-key-regulatory-genes-driving-cell-differentiation">
<h3>Identifying key regulatory genes driving cell differentiation<a class="headerlink" href="#identifying-key-regulatory-genes-driving-cell-differentiation" title="Link to this heading">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">Endocardium_gene_expression_matrix</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;examples/data/zebrafish/additional_input_data/Exp_Endocardium.csv&#39;</span><span class="p">,</span><span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">Epicardium_gene_expression_matrix</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;examples/data/zebrafish/additional_input_data/Exp_Epicardium.csv&#39;</span><span class="p">,</span><span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">Monocytes_gene_expression_matrix</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;examples/data/zebrafish/additional_input_data/Exp_Monocytes.csv&#39;</span><span class="p">,</span><span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">Fibroblasts_gene_expression_matrix</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;examples/data/zebrafish/additional_input_data/Exp_Fibroblasts.csv&#39;</span><span class="p">,</span><span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">Smoothmusclecells_gene_expression_matrix</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;examples/data/zebrafish/additional_input_data/Exp_muscle.csv&#39;</span><span class="p">,</span><span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">Macrophages_gene_expression_matrix</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;examples/data/zebrafish/additional_input_data/Exp_Macrophages.csv&#39;</span><span class="p">,</span><span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">Tcells_gene_expression_matrix</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;examples/data/zebrafish/additional_input_data/Exp_tcell.csv&#39;</span><span class="p">,</span><span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="n">Endocardium_gene_expression_matrix</span><span class="o">.</span><span class="n">index</span><span class="o">=</span><span class="n">Epicardium_gene_expression_matrix</span><span class="o">.</span><span class="n">index</span><span class="o">=</span><span class="n">Monocytes_gene_expression_matrix</span><span class="o">.</span><span class="n">index</span><span class="o">=</span><span class="n">Fibroblasts_gene_expression_matrix</span><span class="o">.</span><span class="n">index</span><span class="o">=</span><span class="n">Smoothmusclecells_gene_expression_matrix</span><span class="o">.</span><span class="n">index</span><span class="o">=</span><span class="n">Macrophages_gene_expression_matrix</span><span class="o">.</span><span class="n">index</span><span class="o">=</span><span class="n">Tcells_gene_expression_matrix</span><span class="o">.</span><span class="n">index</span><span class="o">=</span><span class="n">target_gene_names</span><span class="o">+</span><span class="n">regulator_names</span>

<span class="n">Endocardium_genes</span><span class="o">=</span><span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">get_high_expression_genes</span><span class="p">(</span><span class="n">Endocardium_gene_expression_matrix</span><span class="p">)</span> <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span><span class="p">(</span><span class="n">target_gene_names</span><span class="p">)]</span>
<span class="n">Epicardium_genes</span><span class="o">=</span><span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">get_high_expression_genes</span><span class="p">(</span><span class="n">Epicardium_gene_expression_matrix</span><span class="p">)</span> <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span><span class="p">(</span><span class="n">target_gene_names</span><span class="p">)]</span>
<span class="n">Monocytes_genes</span><span class="o">=</span><span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">get_high_expression_genes</span><span class="p">(</span><span class="n">Monocytes_gene_expression_matrix</span><span class="p">)</span> <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span><span class="p">(</span><span class="n">target_gene_names</span><span class="p">)]</span>
<span class="n">Fibroblasts_genes</span><span class="o">=</span><span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">get_high_expression_genes</span><span class="p">(</span><span class="n">Fibroblasts_gene_expression_matrix</span><span class="p">)</span> <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span><span class="p">(</span><span class="n">target_gene_names</span><span class="p">)]</span>
<span class="n">Smoothmusclecells_genes</span><span class="o">=</span><span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">get_high_expression_genes</span><span class="p">(</span><span class="n">Smoothmusclecells_gene_expression_matrix</span><span class="p">)</span> <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span><span class="p">(</span><span class="n">target_gene_names</span><span class="p">)]</span>
<span class="n">Macrophages_genes</span><span class="o">=</span><span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">get_high_expression_genes</span><span class="p">(</span><span class="n">Macrophages_gene_expression_matrix</span><span class="p">)</span> <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span><span class="p">(</span><span class="n">target_gene_names</span><span class="p">)]</span>
<span class="n">Tcells_genes</span><span class="o">=</span><span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">get_high_expression_genes</span><span class="p">(</span><span class="n">Tcells_gene_expression_matrix</span><span class="p">)</span> <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span><span class="p">(</span><span class="n">target_gene_names</span><span class="p">)]</span>

<span class="n">high_expression_gene_dict</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;Endocardium&#39;</span><span class="p">:</span><span class="n">Endocardium_genes</span><span class="p">,</span><span class="s1">&#39;Epicardium&#39;</span><span class="p">:</span><span class="n">Epicardium_genes</span><span class="p">,</span><span class="s1">&#39;Monocytes&#39;</span><span class="p">:</span><span class="n">Monocytes_genes</span><span class="p">,</span><span class="s1">&#39;Fibroblasts&#39;</span><span class="p">:</span><span class="n">Fibroblasts_genes</span><span class="p">,</span><span class="s1">&#39;Smoothmusclecells&#39;</span><span class="p">:</span><span class="n">Smoothmusclecells_genes</span><span class="p">,</span><span class="s1">&#39;Macrophages&#39;</span><span class="p">:</span><span class="n">Macrophages_genes</span><span class="p">,</span><span class="s1">&#39;Tcells&#39;</span><span class="p">:</span><span class="n">Tcells_genes</span><span class="p">,</span><span class="s1">&#39;IP1&#39;</span><span class="p">:</span><span class="kc">None</span><span class="p">,</span><span class="s1">&#39;IP2&#39;</span><span class="p">:</span><span class="kc">None</span><span class="p">,</span><span class="s1">&#39;IP3&#39;</span><span class="p">:</span><span class="kc">None</span><span class="p">,</span><span class="s1">&#39;IP4&#39;</span><span class="p">:</span><span class="kc">None</span><span class="p">,</span><span class="s1">&#39;IP5&#39;</span><span class="p">:</span><span class="kc">None</span><span class="p">,</span><span class="s1">&#39;IP6&#39;</span><span class="p">:</span><span class="kc">None</span><span class="p">}</span>
<span class="n">fate_map</span><span class="o">=</span> <span class="n">FateMap</span><span class="p">(</span><span class="n">edges</span><span class="p">,</span><span class="n">high_expression_gene_dict</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">key_regulators_df_concat</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

<span class="n">nodes</span><span class="o">=</span><span class="n">fate_map</span><span class="o">.</span><span class="n">node_internals</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">+</span><span class="p">[</span><span class="n">fate_map</span><span class="o">.</span><span class="n">node_root</span><span class="p">]</span>
<span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">nodes</span><span class="p">:</span>
    <span class="n">key_regulators</span><span class="o">=</span><span class="n">find_diff_genes</span><span class="p">(</span><span class="n">saved_dir</span><span class="p">,</span><span class="n">fate_map</span><span class="p">,</span><span class="mf">0.11</span><span class="p">,</span><span class="n">regulator_names</span><span class="p">,</span><span class="n">target_gene_names</span><span class="p">,</span><span class="n">node</span><span class="p">)</span>
    <span class="n">high_expression_genes_of_ancestor_node</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">key_regulators</span><span class="o">.</span><span class="n">loc</span><span class="p">[((</span><span class="n">key_regulators</span><span class="p">[</span><span class="s1">&#39;PRS&#39;</span><span class="p">]</span><span class="o">&gt;</span><span class="mf">0.35</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">key_regulators</span><span class="p">[</span><span class="s1">&#39;PRN&#39;</span><span class="p">]</span><span class="o">&gt;</span><span class="mf">0.6</span><span class="p">)),</span><span class="s1">&#39;regulator_id&#39;</span><span class="p">])</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span><span class="n">target_gene_names</span><span class="p">))</span>
    <span class="n">high_expression_gene_dict</span><span class="p">[</span><span class="n">node</span><span class="p">]</span><span class="o">=</span><span class="n">high_expression_genes_of_ancestor_node</span>
    <span class="n">fate_map</span> <span class="o">=</span> <span class="n">FateMap</span><span class="p">(</span><span class="n">edges</span><span class="p">,</span><span class="n">high_expression_gene_dict</span><span class="p">)</span>
    <span class="n">key_regulators_df_concat</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">key_regulators_df_concat</span><span class="p">,</span><span class="n">key_regulators</span><span class="p">])</span>
</pre></div>
</div>
<section id="id1">
<h4>Visualization<a class="headerlink" href="#id1" title="Link to this heading">¶</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">nodes</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">key_regulators_df_concat</span><span class="p">[</span><span class="s1">&#39;node_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">())</span>
<span class="n">key_regulators_df_concat_V2</span><span class="o">=</span> <span class="n">key_regulators_df_concat</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="n">plot_diff_genes</span><span class="p">(</span><span class="n">key_regulators_df_concat_V2</span><span class="p">,</span><span class="n">nodes</span><span class="p">,</span><span class="n">regulator_names</span><span class="p">,</span><span class="n">output_path</span><span class="p">)</span>
</pre></div>
</div>
<p><img alt="" src="../_images/10_key_regulator_zebrafish.png" /></p>
</section>
</section>
<section id="identifying-key-regulatory-genes-driving-cell-fate-bias">
<h3>Identifying key regulatory genes driving cell fate bias<a class="headerlink" href="#identifying-key-regulatory-genes-driving-cell-fate-bias" title="Link to this heading">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">ancestor_node_id</span><span class="o">=</span><span class="s1">&#39;IP3&#39;</span>
<span class="n">child_nodes</span> <span class="o">=</span> <span class="p">[</span><span class="n">fate_map</span><span class="o">.</span><span class="n">nodes</span><span class="p">[</span><span class="n">ancestor_node_id</span><span class="p">]</span><span class="o">.</span><span class="n">directed_edges</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">end</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">)]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">child_nodes</span><span class="p">)</span>
<span class="n">high_expression_target_genes_in_child_1</span><span class="o">=</span><span class="n">fate_map</span><span class="o">.</span><span class="n">nodes</span><span class="p">[</span><span class="n">child_nodes</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">high_expression_genes_in_leaf</span>
<span class="n">high_expression_target_genes_in_child_2</span><span class="o">=</span><span class="n">fate_map</span><span class="o">.</span><span class="n">nodes</span><span class="p">[</span><span class="n">child_nodes</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span><span class="o">.</span><span class="n">high_expression_genes_in_leaf</span>
 
<span class="n">Tar_1</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">target_gene_names</span><span class="p">)</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span><span class="n">high_expression_target_genes_in_child_1</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">high_expression_target_genes_in_child_2</span><span class="p">))</span>
<span class="n">Tar_2</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">target_gene_names</span><span class="p">)</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span><span class="n">high_expression_target_genes_in_child_2</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">high_expression_target_genes_in_child_1</span><span class="p">))</span>


<span class="n">key_regulators_df_IP3</span><span class="o">=</span><span class="n">find_fate_bias_genes</span><span class="p">(</span><span class="n">dynamic_networks_dict</span><span class="p">,</span> <span class="s1">&#39;IP3&#39;</span><span class="p">,</span><span class="n">Tar_1</span><span class="p">,</span> <span class="n">Tar_2</span><span class="p">,</span> <span class="n">regulator_names</span><span class="p">)</span>
</pre></div>
</div>
<section id="id2">
<h4>Visualization<a class="headerlink" href="#id2" title="Link to this heading">¶</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">df1</span> <span class="o">=</span> <span class="n">key_regulators_df_IP3</span><span class="p">[</span><span class="mi">0</span><span class="p">][[</span><span class="s1">&#39;Regulatory gene&#39;</span><span class="p">,</span> <span class="s1">&#39;PosRegStrBias&#39;</span><span class="p">,</span> <span class="s1">&#39;RegBias&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">df2</span> <span class="o">=</span> <span class="n">key_regulators_df_IP3</span><span class="p">[</span><span class="mi">1</span><span class="p">][[</span><span class="s1">&#39;Regulatory gene&#39;</span><span class="p">,</span> <span class="s1">&#39;PosRegStrBias&#39;</span><span class="p">,</span> <span class="s1">&#39;RegBias&#39;</span><span class="p">]]</span>
<span class="n">merged_df</span> <span class="o">=</span> <span class="n">df1</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
    <span class="n">df2</span><span class="p">,</span> 
    <span class="n">on</span><span class="o">=</span><span class="s1">&#39;Regulatory gene&#39;</span><span class="p">,</span> 
    <span class="n">how</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">,</span>  
    <span class="n">suffixes</span><span class="o">=</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;_</span><span class="si">{</span><span class="n">child_nodes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;_</span><span class="si">{</span><span class="n">child_nodes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="p">)</span>

<span class="n">plot_fate_bias_genes</span><span class="p">(</span><span class="n">merged_df</span><span class="p">,</span> <span class="n">child_nodes</span><span class="p">,</span><span class="n">output_path</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mf">1.5</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
</pre></div>
</div>
<p><img alt="" src="../_images/fate_bias_zebrafish.png" /></p>
</section>
</section>
<section id="identifying-specific-and-constitutive-regulatory-interactions">
<h3>Identifying specific and constitutive regulatory interactions.<a class="headerlink" href="#identifying-specific-and-constitutive-regulatory-interactions" title="Link to this heading">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">X</span><span class="p">,</span><span class="n">centers</span><span class="p">,</span><span class="n">weight_matrix</span><span class="o">=</span><span class="n">run_fuzzy_C_means_clustering</span><span class="p">(</span><span class="n">saved_dir</span><span class="p">,</span><span class="n">fate_map</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="n">regulator_names</span><span class="p">,</span> <span class="n">target_gene_names</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">regulator_names</span><span class="p">),</span><span class="nb">len</span><span class="p">(</span><span class="n">target_gene_names</span><span class="p">),</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="n">edges_cluster_to_nodes</span><span class="o">=</span><span class="n">map_edge_clusters_to_nodes</span><span class="p">(</span><span class="n">saved_dir</span><span class="p">,</span> <span class="n">fate_map</span><span class="p">,</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">regulator_names</span><span class="p">,</span><span class="n">target_gene_names</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mf">0.4</span><span class="p">,</span><span class="n">X</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">regulator_names</span><span class="p">),</span><span class="nb">len</span><span class="p">(</span><span class="n">target_gene_names</span><span class="p">),</span><span class="n">weight_matrix</span><span class="p">)</span>
</pre></div>
</div>
<section id="id3">
<h4>Visualization<a class="headerlink" href="#id3" title="Link to this heading">¶</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">plot_edge_cluster_weight</span><span class="p">(</span><span class="n">weight_matrix</span><span class="p">,</span><span class="n">output_path</span><span class="p">)</span>
</pre></div>
</div>
<p><img alt="" src="../_images/11_edge_cluster_weight_zebrafish.png" /></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">plot_edge_cluster_to_nodes</span><span class="p">(</span><span class="n">edges_cluster_to_nodes</span><span class="p">,</span> <span class="n">output_path</span><span class="p">)</span>
</pre></div>
</div>
<p><img alt="" src="../_images/12_edge_cluster_node_zebrafish.png" /></p>
</section>
</section>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="../API/index.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">API</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="02_Realistic_data.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Example of Caenorhabditis elegans dataset (Full)</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2024, Xiaochen Yu
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Example of Zebrafish</a><ul>
<li><a class="reference internal" href="#data-preprocessing">Data Preprocessing</a><ul>
<li><a class="reference internal" href="#input-data">1. Input Data</a></li>
<li><a class="reference internal" href="#sequence-alignment">2. Sequence Alignment</a></li>
<li><a class="reference internal" href="#data-parsing-and-processing">3. Data Parsing and Processing</a></li>
<li><a class="reference internal" href="#mutation-matrix-generation">4. Mutation Matrix Generation</a></li>
<li><a class="reference internal" href="#example-output-files">Example Output Files</a></li>
</ul>
</li>
<li><a class="reference internal" href="#time-scaled-cell-fate-map-reconstruction">Time-scaled cell fate map reconstruction</a></li>
<li><a class="reference internal" href="#gene-regulatory-network-inference">Gene regulatory network inference</a><ul>
<li><a class="reference internal" href="#input-single-cell-gene-expression-and-co-accessibility-profiles">Input single-cell gene expression and co-accessibility profiles</a></li>
<li><a class="reference internal" href="#run-the-gene-regulatory-network-inference-function">Run the gene regulatory network inference function</a></li>
</ul>
</li>
<li><a class="reference internal" href="#downstream-analysis">Downstream analysis</a><ul>
<li><a class="reference internal" href="#revealing-how-inferred-grns-reconfigure-along-cell-lineages">Revealing how inferred GRNs reconfigure along cell lineages</a><ul>
<li><a class="reference internal" href="#visualization">Visualization</a></li>
</ul>
</li>
<li><a class="reference internal" href="#identifying-key-regulatory-genes-driving-cell-differentiation">Identifying key regulatory genes driving cell differentiation</a><ul>
<li><a class="reference internal" href="#id1">Visualization</a></li>
</ul>
</li>
<li><a class="reference internal" href="#identifying-key-regulatory-genes-driving-cell-fate-bias">Identifying key regulatory genes driving cell fate bias</a><ul>
<li><a class="reference internal" href="#id2">Visualization</a></li>
</ul>
</li>
<li><a class="reference internal" href="#identifying-specific-and-constitutive-regulatory-interactions">Identifying specific and constitutive regulatory interactions.</a><ul>
<li><a class="reference internal" href="#id3">Visualization</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="../_static/documentation_options.js?v=d45e8c67"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/scripts/furo.js?v=5fa4622c"></script>
    </body>
</html>